
<html>
<head>
<meta http-equiv="CONTENT-TYPE" content="text/html; charset=utf-8" />
<title>Compara Schema Documentation</title>

<script language="Javascript" type="text/javascript">
  var img_plus   = '<img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>';
  var img_minus  = '<img src="/i/16/minus-button.png" style="width:12px;height:12px;vertical-align:middle;margin-right:6px" alt="hide"/>';
  var link_text  = 'columns';
  var span_open  = ' <span style="vertical-align:middle">';
  var span_close = '</span>';

  // Function to show/hide the columns table
  function show_hide (param, type) {
    // Example tables
    if (type === 'example') {
      div    = document.getElementById('ex_'+param);
      alink  = document.getElementById('e_'+param);
      a_text = 'query results'; 
    } 
    // Schema tables
    else { 
      div    = document.getElementById('div_'+param);
      alink  = document.getElementById('a_'+param);
      a_text = link_text;
    }  
    
    if (div.style.display=='inline') {
      div.style.display='none';
      alink.innerHTML=img_plus+span_open+'Show '+a_text+span_close;
    }
    else {
      if (div.style.display=='none') {
        div.style.display='inline';
        alink.innerHTML=img_minus+span_open+'Hide '+a_text+span_close;
      }
    }
  }
  
  // Function to show/hide all the tables
  function show_hide_all () {
    expand_flag = document.getElementById('expand');
    divs = document.getElementsByTagName('div');
    for(var i=0; i < divs.length; i++){
      param = divs[i].id.substring(4);
      div   = document.getElementById('div_'+param);
      alink = document.getElementById('a_'+param);
      
      if (div && alink) {
        if (expand_flag.value=='0') {
          div.style.display='inline';
          alink.innerHTML=img_minus+span_open+'Hide '+link_text+span_close;
        }
        else {
          div.style.display='none';
          alink.innerHTML=img_plus+span_open+'Show '+link_text+span_close;
        }
      }
    }
    if (expand_flag.value=='0') {
      expand_flag.value='1';
    }
    else {
      expand_flag.value='0';
    }
  }
</script>

</head>
<body>

<h1>Ensembl Compara Schema Documentation</h1>
<h2>Introduction</h2>
<p>
This document describes the tables that make up the Ensembl Compara schema. Tables are listed grouped in different categories, and the purpose of each table is explained. Several examples are also given. They are intended to allow people to familiarise themselves with the schema.
</p>

<p>
Two (partially overlapping) parts of Compara schema diagram are
available: <a target="_blank" href="./diagrams/genomic_schema.png">Genomic alignments</a> and <a target="_blank" href="./diagrams/protein_schema.png">Gene trees and homologies</a>. The combined diagram can be found <a target="_blank" href="./diagrams/compara_schema.png">here</a>
</p>



<h3 id="top">List of the tables:</h3>
<div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#3CB371;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">General Tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#meta"><b>meta</b></a></li>
        <li><a href="#ncbi_taxa_node"><b>ncbi_taxa_node</b></a></li>
        <li><a href="#ncbi_taxa_name"><b>ncbi_taxa_name</b></a></li>
        <li><a href="#genome_db"><b>genome_db</b></a></li>
        <li><a href="#species_set_header"><b>species_set_header</b></a></li>
        <li><a href="#species_set"><b>species_set</b></a></li>
        <li><a href="#species_set_tag"><b>species_set_tag</b></a></li>
        <li><a href="#method_link"><b>method_link</b></a></li>
        <li><a href="#method_link_species_set"><b>method_link_species_set</b></a></li>
        <li><a href="#method_link_species_set_tag"><b>method_link_species_set_tag</b></a></li>
        <li><a href="#species_tree_node"><b>species_tree_node</b></a></li>
        <li><a href="#species_tree_root"><b>species_tree_root</b></a></li>
        <li><a href="#species_tree_node_tag"><b>species_tree_node_tag</b></a></li>

      </ul>
    </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px;width:200px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#FF8500;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Genomic alignments tables</h2>
    </div>
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#synteny_region"><b>synteny_region</b></a></li>
        <li><a href="#dnafrag"><b>dnafrag</b></a></li>
        <li><a href="#dnafrag_region"><b>dnafrag_region</b></a></li>
        <li><a href="#genomic_align_block"><b>genomic_align_block</b></a></li>
        <li><a href="#genomic_align_tree"><b>genomic_align_tree</b></a></li>
        <li><a href="#genomic_align"><b>genomic_align</b></a></li>
        <li><a href="#conservation_score"><b>conservation_score</b></a></li>
        <li><a href="#constrained_element"><b>constrained_element</b></a></li>

      </ul>
    </div>

  <div style="background-color:#F2F2F2;border-radius:5px;margin-bottom:15px;float:left;margin-right:20px">
    <div style="padding:2px 5px;background-color:#336;border-top-left-radius:5px;border-top-right-radius:5px">
      <div style="background-color:#1E90FF;border:1px solid #FFF;padding:0px 8px;display:inline;vertical-align:middle"></div>
      <h2 style="margin-left:8px;display:inline;color:#FFF;vertical-align:middle">Gene trees and homologies tables</h2>
    </div>
      <div style="float:left">
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#sequence"><b>sequence</b></a></li>
        <li><a href="#gene_member"><b>gene_member</b></a></li>
        <li><a href="#seq_member"><b>seq_member</b></a></li>
        <li><a href="#external_db"><b>external_db</b></a></li>
        <li><a href="#member_xref"><b>member_xref</b></a></li>
        <li><a href="#other_member_sequence"><b>other_member_sequence</b></a></li>
        <li><a href="#peptide_align_feature"><b>peptide_align_feature</b></a></li>
        <li><a href="#family"><b>family</b></a></li>
        <li><a href="#family_member"><b>family_member</b></a></li>
        <li><a href="#gene_align"><b>gene_align</b></a></li>
        <li><a href="#gene_align_member"><b>gene_align_member</b></a></li>
        <li><a href="#gene_tree_node"><b>gene_tree_node</b></a></li>
        <li><a href="#gene_tree_root"><b>gene_tree_root</b></a></li>
        <li><a href="#gene_tree_node_tag"><b>gene_tree_node_tag</b></a></li>
        <li><a href="#gene_tree_root_tag"><b>gene_tree_root_tag</b></a></li>

      </ul>
      </div>
      <div style="float:left">
      <ul style="padding:0px 4px 0px 22px;margin-bottom:2px">
        <li><a href="#gene_tree_node_attr"><b>gene_tree_node_attr</b></a></li>
        <li><a href="#hmm_profile"><b>hmm_profile</b></a></li>
        <li><a href="#hmm_annot"><b>hmm_annot</b></a></li>
        <li><a href="#hmm_curated_annot"><b>hmm_curated_annot</b></a></li>
        <li><a href="#homology"><b>homology</b></a></li>
        <li><a href="#homology_member"><b>homology_member</b></a></li>
        <li><a href="#mapping_session"><b>mapping_session</b></a></li>
        <li><a href="#stable_id_history"><b>stable_id_history</b></a></li>
        <li><a href="#CAFE_gene_family"><b>CAFE_gene_family</b></a></li>
        <li><a href="#CAFE_species_gene"><b>CAFE_species_gene</b></a></li>

      </ul>
      </div>
    </div>

  <div style="clear:both" />
</div>
  <input type="button" onclick="show_hide_all()" class="fbutton" value="Show/hide all"/>
  <input type="hidden" id="expand" value="0" />
  
<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #3CB371">
  <h2 style="display:inline;color:#000">General Tables</h2>
</div>
<p style="width:800px">These are general tables used in the Compara schema</p>
  <div id="meta" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">meta</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_meta" style="cursor:pointer;text-decoration:none" onclick="show_hide('meta')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores meta information about the compara database</p>

  <div id="div_meta" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>meta_id</b></td><td>INT</td><td></td><td>Internal unique ID for the table</td><td>primary   key</td></tr>
      <tr class="bg2"><td><b>species_id</b></td><td>INT</td><td>1</td><td>Only used in core databases</td><td>unique    key: species_key_value_idx<br />key: species_value_idx</td></tr>
      <tr class="bg1"><td><b>meta_key</b></td><td>VARCHAR(40)</td><td></td><td>Key for the key/value pair</td><td>unique    key: species_key_value_idx</td></tr>
      <tr class="bg2"><td><b>meta_value</b></td><td>TEXT</td><td></td><td>Value for the key/value pair</td><td>unique    key: species_key_value_idx<br />key: species_value_idx</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query defines which API version must be used to access this database. </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM meta WHERE meta_key = "schema_version";</pre> <a id="e_meta" style="cursor:pointer;text-decoration:none" onclick="show_hide('meta','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_meta" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>meta_id</th><th>species_id</th><th>meta_key</th><th>meta_value</th></tr>      <tr><td>1</td><td></td><td>schema_version</td><td>82</td></tr>    </table>
  </div></div>
  <div id="ncbi_taxa_node" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">ncbi_taxa_node</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_ncbi_taxa_node" style="cursor:pointer;text-decoration:none" onclick="show_hide('ncbi_taxa_node')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains all taxa used in this database, which mirror the data and tree structure from NCBI Taxonomy database (for more details see ensembl-compara/script/taxonomy/README-taxonomy which explain our import process)</p>

  <div id="div_ncbi_taxa_node" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>taxon_id</b></td><td>int(10)</td><td></td><td>The NCBI Taxonomy ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>parent_id</b></td><td>int(10)</td><td></td><td>The parent taxonomy ID for this node (refers to ncbi_taxa_node.taxon_id)</td><td>key: parent_id</td></tr>
      <tr class="bg1"><td><b>rank</b></td><td>char(32)</td><td>''</td><td>E.g. kingdom, family, genus, etc.</td><td>key: rank</td></tr>
      <tr class="bg2"><td><b>genbank_hidden_flag</b></td><td>tinyint(1)</td><td>0</td><td>Boolean value which defines whether this rank is used or not in the abbreviated lineage</td><td></td></tr>
      <tr class="bg1"><td><b>left_index</b></td><td>int(10)</td><td>0</td><td>Sub-set left index. All sub-nodes have left_index and right_index values larger than this left_index</td><td>key: left_index</td></tr>
      <tr class="bg2"><td><b>right_index</b></td><td>int(10)</td><td>0</td><td>Sub-set right index. All sub-nodes have left_index and right_index values smaller than this right_index</td><td>key: right_index</td></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>int(10)</td><td>1</td><td>The root taxonomy ID for this node (refers to ncbi_taxa_node.taxon_id)</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This examples shows how to get the lineage for Homo sapiens: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT n2.taxon_id, n2.parent_id, na.name, n2.rank, n2.left_index, n2.right_index FROM ncbi_taxa_node n1 JOIN (ncbi_taxa_node n2 LEFT JOIN ncbi_taxa_name na ON n2.taxon_id = na.taxon_id AND na.name_class = "scientific name")  ON n2.left_index <= n1.left_index AND n2.right_index >= n1.right_index WHERE n1.taxon_id = 9606 ORDER BY left_index;</pre> <a id="e_ncbi_taxa_node" style="cursor:pointer;text-decoration:none" onclick="show_hide('ncbi_taxa_node','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_ncbi_taxa_node" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>n2.taxon_id</th><th>n2.parent_id</th><th>na.name</th><th>n2.rank</th><th>n2.left_index</th><th>n2.right_index</th></tr>      <tr><td>1</td><td>0</td><td>root</td><td>no rank</td><td>1</td><td>2638750</td></tr>      <tr class="bg2"><td>131567</td><td>1</td><td>cellular organisms</td><td>no rank</td><td>2</td><td>2339569</td></tr>      <tr><td>2759</td><td>131567</td><td>Eukaryota</td><td>superkingdom</td><td>21747</td><td>1535624</td></tr>      <tr class="bg2"><td>33154</td><td>2759</td><td>Opisthokonta</td><td>no rank</td><td>360752</td><td>1459587</td></tr>      <tr><td>33208</td><td>33154</td><td>Metazoa</td><td>kingdom</td><td>361039</td><td>1212656</td></tr>      <tr class="bg2"><td>6072</td><td>33208</td><td>Eumetazoa</td><td>no rank</td><td>361310</td><td>1207105</td></tr>      <tr><td>33213</td><td>6072</td><td>Bilateria</td><td>no rank</td><td>361571</td><td>1192602</td></tr>      <tr class="bg2"><td>33511</td><td>33213</td><td>Deuterostomia</td><td>no rank</td><td>1021640</td><td>1177563</td></tr>      <tr><td>7711</td><td>33511</td><td>Chordata</td><td>phylum</td><td>1028631</td><td>1177562</td></tr>      <tr class="bg2"><td>89593</td><td>7711</td><td>Craniata</td><td>subphylum</td><td>1028636</td><td>1176457</td></tr>      <tr><td>7742</td><td>89593</td><td>Vertebrata</td><td>no rank</td><td>1028637</td><td>1176456</td></tr>      <tr class="bg2"><td>7776</td><td>7742</td><td>Gnathostomata</td><td>no rank</td><td>1028642</td><td>1176169</td></tr>      <tr><td>117570</td><td>7776</td><td>Teleostomi</td><td>no rank</td><td>1028643</td><td>1173014</td></tr>      <tr class="bg2"><td>117571</td><td>117570</td><td>Euteleostomi</td><td>no rank</td><td>1028644</td><td>1173013</td></tr>      <tr><td>8287</td><td>117571</td><td>Sarcopterygii</td><td>no rank</td><td>1028645</td><td>1112708</td></tr>      <tr class="bg2"><td>1338369</td><td>8287</td><td>Dipnotetrapodomorpha</td><td>no rank</td><td>1028658</td><td>1112707</td></tr>      <tr><td>32523</td><td>1338369</td><td>Tetrapoda</td><td>no rank</td><td>1028711</td><td>1112706</td></tr>      <tr class="bg2"><td>32524</td><td>32523</td><td>Amniota</td><td>no rank</td><td>1028712</td><td>1096581</td></tr>      <tr><td>40674</td><td>32524</td><td>Mammalia</td><td>class</td><td>1077149</td><td>1096580</td></tr>      <tr class="bg2"><td>32525</td><td>40674</td><td>Theria</td><td>no rank</td><td>1077176</td><td>1096569</td></tr>      <tr><td>9347</td><td>32525</td><td>Eutheria</td><td>no rank</td><td>1077177</td><td>1095470</td></tr>      <tr class="bg2"><td>1437010</td><td>9347</td><td>Boreoeutheria</td><td>no rank</td><td>1077186</td><td>1095001</td></tr>      <tr><td>314146</td><td>1437010</td><td>Euarchontoglires</td><td>superorder</td><td>1077187</td><td>1086616</td></tr>      <tr class="bg2"><td>9443</td><td>314146</td><td>Primates</td><td>order</td><td>1084948</td><td>1086615</td></tr>      <tr><td>376913</td><td>9443</td><td>Haplorrhini</td><td>suborder</td><td>1084949</td><td>1086156</td></tr>      <tr class="bg2"><td>314293</td><td>376913</td><td>Simiiformes</td><td>infraorder</td><td>1084976</td><td>1086155</td></tr>      <tr><td>9526</td><td>314293</td><td>Catarrhini</td><td>parvorder</td><td>1084977</td><td>1085684</td></tr>      <tr class="bg2"><td>314295</td><td>9526</td><td>Hominoidea</td><td>superfamily</td><td>1085538</td><td>1085683</td></tr>      <tr><td>9604</td><td>314295</td><td>Hominidae</td><td>family</td><td>1085539</td><td>1085600</td></tr>      <tr class="bg2"><td>207598</td><td>9604</td><td>Homininae</td><td>subfamily</td><td>1085554</td><td>1085599</td></tr>      <tr><td>9605</td><td>207598</td><td>Homo</td><td>genus</td><td>1085573</td><td>1085582</td></tr>      <tr class="bg2"><td>9606</td><td>9605</td><td>Homo sapiens</td><td>species</td><td>1085574</td><td>1085579</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#ncbi_taxa_name">ncbi_taxa_name</a></li>
  </ul>


  <div id="ncbi_taxa_name" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">ncbi_taxa_name</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_ncbi_taxa_name" style="cursor:pointer;text-decoration:none" onclick="show_hide('ncbi_taxa_name')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains different names, aliases and meta data for the taxa used in Ensembl.</p>

  <div id="div_ncbi_taxa_name" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>taxon_id</b></td><td>int(10)</td><td></td><td>External reference to taxon_id in <a href="#ncbi_taxa_node">ncbi_taxa_node</a></td><td>key: taxon_id</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td></td><td>Information assigned to this taxon_id</td><td>key: name</td></tr>
      <tr class="bg1"><td><b>name_class</b></td><td>varchar(50)</td><td></td><td>Type of information. e.g. common name, genbank_synonym, scientif name, etc.</td><td>key: name_class</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">Here is an example on how to get the taxonomic ID for a species: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM ncbi_taxa_name WHERE name_class = "scientific name" AND name = "Homo sapiens";</pre> <a id="e_ncbi_taxa_name" style="cursor:pointer;text-decoration:none" onclick="show_hide('ncbi_taxa_name','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_ncbi_taxa_name" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>taxon_id</th><th>name</th><th>name_class</th></tr>      <tr><td>9606</td><td>Homo sapiens</td><td>scientific name</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#ncbi_taxa_node">ncbi_taxa_node</a></li>
  </ul>


  <div id="genome_db" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">genome_db</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_genome_db" style="cursor:pointer;text-decoration:none" onclick="show_hide('genome_db')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains information about the version of the genome assemblies used in this database</p>

  <div id="div_genome_db" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>genome_db_id</b></td><td>int(10)</td><td></td><td>Internal unique ID for this table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>taxon_id</b></td><td>int(10)</td><td>NULL</td><td>External reference to taxon_id in the <a href="#ncbi_taxa_node">ncbi_taxa_node</a> table</td><td></td></tr>
      <tr class="bg1"><td><b>name</b></td><td>varchar(128)</td><td>''</td><td>Species name</td><td>unique: name</td></tr>
      <tr class="bg2"><td><b>assembly</b></td><td>varchar(100)</td><td>''</td><td>Assembly version of the genome</td><td>unique: name</td></tr>
      <tr class="bg1"><td><b>genebuild</b></td><td>varchar(100)</td><td>''</td><td>Version of the genebuild</td><td></td></tr>
      <tr class="bg2"><td><b>has_karyotype</b></td><td>tinyint(1)</td><td>0</td><td>Whether the genome has a karyotype</td><td></td></tr>
      <tr class="bg1"><td><b>is_high_coverage</b></td><td>tinyint(1)</td><td>0</td><td>Whether the assembly coverage depth is high enough</td><td></td></tr>
      <tr class="bg2"><td><b>genome_component</b></td><td>varchar(5)</td><td>NULL</td><td>Only used for polyploid genomes: the name of the genome component</td><td>unique: name</td></tr>
      <tr class="bg1"><td><b>locator</b></td><td>varchar(400)</td><td></td><td>Used for production purposes or for user configuration in in-house installation.</td><td></td></tr>
      <tr class="bg2"><td><b>first_release</b></td><td>smallint</td><td></td><td>The first release this genome was present in</td><td></td></tr>
      <tr class="bg1"><td><b>last_release</b></td><td>smallint</td><td></td><td>The last release this genome was present in, or NULL if it is still current</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query shows the entries for human and chicken </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM genome_db WHERE name IN ("Homo_sapiens", "Gallus_gallus");</pre> <a id="e_genome_db" style="cursor:pointer;text-decoration:none" onclick="show_hide('genome_db','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_genome_db" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>genome_db_id</th><th>taxon_id</th><th>name</th><th>assembly</th><th>genebuild</th><th>has_karyotype</th><th>is_high_coverage</th><th>genome_component</th><th>locator</th><th>first_release</th><th>last_release</th></tr>      <tr><td>142</td><td>9031</td><td>gallus_gallus</td><td>Galgal4</td><td>2011-12-Ensembl</td><td>1</td><td>1</td><td></td><td></td><td>71</td><td></td></tr>      <tr class="bg2"><td>150</td><td>9606</td><td>homo_sapiens</td><td>GRCh38</td><td>2014-01-Ensembl</td><td>1</td><td>1</td><td></td><td></td><td>76</td><td></td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#ncbi_taxa_node">ncbi_taxa_node</a></li>
  </ul>


  <div id="species_set_header" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">species_set_header</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_species_set_header" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_set_header')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Header for the <a href="#species_set">species_set</a> table which groups or sets of species which are used in the <a href="#method_link_species_set">method_link_species_set</a> table.</p>

  <div id="div_species_set_header" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>species_set_id</b></td><td>int(10)</td><td></td><td>Internal (non-unique) ID for the table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td>''</td><td>Name of the species set (e.g. "H.sap" for one species, "H.sap-M.mus" for two species and "amniotes" for larger sets)</td><td></td></tr>
      <tr class="bg1"><td><b>size</b></td><td>int(10)</td><td></td><td>Number of species in the set</td><td></td></tr>
      <tr class="bg2"><td><b>first_release</b></td><td>smallint</td><td></td><td>The first release this set genome was present in</td><td></td></tr>
      <tr class="bg1"><td><b>last_release</b></td><td>smallint</td><td></td><td>The last release this set was present in, or NULL if it is still current</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query shows the first 10 species_sets having human </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT ssh.species_set_id, ssh.name, ssh.size, gdb.genome_db_id, gdb.name, gdb.assembly FROM species_set_header ssh JOIN species_set ss USING(species_set_id) JOIN genome_db gdb USING(genome_db_id) WHERE ssh.name="primates";  </pre> <a id="e_species_set_header" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_set_header','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_species_set_header" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>ssh.species_set_id</th><th>ssh.name</th><th>ssh.size</th><th>gdb.genome_db_id</th><th>gdb.name</th><th>gdb.assembly</th></tr>      <tr><td>35886</td><td>primates</td><td>8</td><td>31</td><td>macaca_mulatta</td><td>MMUL_1</td></tr>      <tr class="bg2"><td>35886</td><td>primates</td><td>8</td><td>60</td><td>pongo_abelii</td><td>PPYG2</td></tr>      <tr><td>35886</td><td>primates</td><td>8</td><td>117</td><td>callithrix_jacchus</td><td>C_jacchus3.2.1</td></tr>      <tr class="bg2"><td>35886</td><td>primates</td><td>8</td><td>123</td><td>gorilla_gorilla</td><td>gorGor3.1</td></tr>      <tr><td>35886</td><td>primates</td><td>8</td><td>125</td><td>pan_troglodytes</td><td>CHIMP2.1.4</td></tr>      <tr class="bg2"><td>35886</td><td>primates</td><td>8</td><td>150</td><td>homo_sapiens</td><td>GRCh38</td></tr>      <tr><td>35886</td><td>primates</td><td>8</td><td>151</td><td>papio_anubis</td><td>PapAnu2.0</td></tr>      <tr class="bg2"><td>35886</td><td>primates</td><td>8</td><td>153</td><td>chlorocebus_sabaeus</td><td>ChlSab1.1</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
    <li><a href="#genome_db">genome_db</a></li>
    <li><a href="#species_set">species_set</a></li>
  </ul>


  <div id="species_set" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">species_set</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_species_set" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_set')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Describes the content of each species-set (<a href="#species_set_header">species_set_header</a>) as a set of <a href="#genome_db">genome_db</a> objects</p>

  <div id="div_species_set" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>species_set_id</b></td><td>int(10)</td><td></td><td>Internal ID for the table, foreign key to <a href="#species_set_header">species_set_header</a></td><td>primary key</td></tr>
      <tr class="bg2"><td><b>genome_db_id</b></td><td>int(10)</td><td></td><td>External reference to genome_db_id in the <a href="#genome_db">genome_db</a> table</td><td>primary key</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query shows the first 10 species_sets having human </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT species_set_id, GROUP_CONCAT(name) AS species FROM species_set JOIN genome_db USING(genome_db_id) GROUP BY species_set_id HAVING species LIKE '%homo_sapiens%' ORDER BY species_set_id LIMIT 10;</pre> <a id="e_species_set" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_set','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_species_set" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>species_set_id</th><th>species</th></tr>      <tr><td>35673</td><td>mus_musculus,homo_sapiens</td></tr>      <tr class="bg2"><td>35674</td><td>choloepus_hoffmanni,homo_sapiens</td></tr>      <tr><td>35675</td><td>nomascus_leucogenys,homo_sapiens</td></tr>      <tr class="bg2"><td>35676</td><td>oryctolagus_cuniculus,homo_sapiens</td></tr>      <tr><td>35677</td><td>cavia_porcellus,homo_sapiens</td></tr>      <tr class="bg2"><td>35678</td><td>otolemur_garnettii,homo_sapiens</td></tr>      <tr><td>35679</td><td>pongo_abelii,homo_sapiens</td></tr>      <tr class="bg2"><td>35680</td><td>dasypus_novemcinctus,homo_sapiens</td></tr>      <tr><td>35681</td><td>dipodomys_ordii,homo_sapiens</td></tr>      <tr class="bg2"><td>35682</td><td>erinaceus_europaeus,homo_sapiens</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
    <li><a href="#genome_db">genome_db</a></li>
    <li><a href="#species_set_header">species_set_header</a></li>
  </ul>


  <div id="species_set_tag" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">species_set_tag</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_species_set_tag" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_set_tag')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains descriptive tags for the species_set_ids in the species_set table. It is used to store options on clades and group of species. It has been initially developed for the gene tree view.</p>

  <div id="div_species_set_tag" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>species_set_id</b></td><td>int(10)</td><td></td><td>External reference to species_set_id in the <a href="#species_set">species_set</a> table</td><td>unique key: tag_species_set_id</td></tr>
      <tr class="bg2"><td><b>tag</b></td><td>varchar(50)</td><td></td><td>Tag name</td><td>unique key: tag_species_set_id</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>mediumtext</td><td></td><td>Tag value</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#species_set">species_set</a></li>
  </ul>


  <div id="method_link" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">method_link</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_method_link" style="cursor:pointer;text-decoration:none" onclick="show_hide('method_link')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table specifies which kind of link can exist between entities in compara (dna/dna alignment, synteny regions, homologous gene pairs, etc...)<br/>NOTE: We use method_link_ids between 1 and 100 for DNA-DNA alignments, between 101 and 200 for genomic syntenies, between 201 and 300 for protein homologies, between 301 and 400 for protein families and between 401 and 500 for protein and ncRNA trees. Each category corresponds to data stored in different tables.</p>

  <div id="div_method_link" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>method_link_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>type</b></td><td>varchar(50)</td><td>''</td><td>The common name of the linking method between species</td><td>unique key: type</td></tr>
      <tr class="bg1"><td><b>class</b></td><td>varchar(50)</td><td>''</td><td>Description of type of data associated with the \"type\" field and the main table to find these data</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">These are our current entries: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM method_link;</pre> <a id="e_method_link" style="cursor:pointer;text-decoration:none" onclick="show_hide('method_link','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_method_link" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>method_link_id</th><th>type</th><th>class</th></tr>      <tr><td>1</td><td>BLASTZ_NET</td><td>GenomicAlignBlock.pairwise_alignment</td></tr>      <tr class="bg2"><td>10</td><td>PECAN</td><td>GenomicAlignBlock.multiple_alignment</td></tr>      <tr><td>11</td><td>GERP_CONSTRAINED_ELEMENT</td><td>ConstrainedElement.constrained_element</td></tr>      <tr class="bg2"><td>13</td><td>EPO</td><td>GenomicAlignTree.ancestral_alignment</td></tr>      <tr><td>14</td><td>EPO_LOW_COVERAGE</td><td>GenomicAlignTree.tree_alignment</td></tr>      <tr class="bg2"><td>16</td><td>LASTZ_NET</td><td>GenomicAlignBlock.pairwise_alignment</td></tr>      <tr><td>19</td><td>LASTZ_PATCH</td><td>GenomicAlignBlock.pairwise_alignment</td></tr>      <tr class="bg2"><td>101</td><td>SYNTENY</td><td>SyntenyRegion.synteny</td></tr>      <tr><td>201</td><td>ENSEMBL_ORTHOLOGUES</td><td>Homology.homology</td></tr>      <tr class="bg2"><td>202</td><td>ENSEMBL_PARALOGUES</td><td>Homology.homology</td></tr>      <tr><td>205</td><td>ENSEMBL_PROJECTIONS</td><td>Homology.homology</td></tr>      <tr class="bg2"><td>301</td><td>FAMILY</td><td>Family.family</td></tr>      <tr><td>401</td><td>PROTEIN_TREES</td><td>ProteinTree.protein_tree_node</td></tr>      <tr class="bg2"><td>402</td><td>NC_TREES</td><td>NCTree.nc_tree_node</td></tr>      <tr><td>501</td><td>GERP_CONSERVATION_SCORE</td><td>ConservationScore.conservation_score</td></tr>      <tr class="bg2"><td>600</td><td>SPECIES_TREE</td><td>SpeciesTree.species_tree_root</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
    <li><a href="#species_set">species_set</a></li>
  </ul>


  <div id="method_link_species_set" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">method_link_species_set</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_method_link_species_set" style="cursor:pointer;text-decoration:none" onclick="show_hide('method_link_species_set')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains information about the comparisons stored in the database. A given method_link_species_set_id exist for each comparison made and relates a method_link_id in <a href="#method_link">method_link</a> with a set of species (species_set_id) in the <a href="#species_set">species_set</a> table.</p>

  <div id="div_method_link_species_set" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>method_link_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_id in the <a href="#method_link">method_link</a> table</td><td>unique key: method_link_id</td></tr>
      <tr class="bg1"><td><b>species_set_id</b></td><td>int(10)</td><td></td><td>External reference to species_set_id in the <a href="#species_set">species_set</a> table</td><td>unique key: method_link_id</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(255)</td><td>''</td><td>Human-readable description for this method_link_species_set</td><td></td></tr>
      <tr class="bg1"><td><b>source</b></td><td>varchar(255)</td><td>'ensembl'</td><td>Source of the data. Currently either "ensembl" or "ucsc" if data were imported from UCSC</td><td></td></tr>
      <tr class="bg2"><td><b>url</b></td><td>varchar(255)</td><td>''</td><td>A URL where you can find the orignal data if they were imported</td><td></td></tr>
      <tr class="bg1"><td><b>first_release</b></td><td>smallint</td><td></td><td>The first release this analysis was present in</td><td></td></tr>
      <tr class="bg2"><td><b>last_release</b></td><td>smallint</td><td></td><td>The last release this analysis was present in, or NULL if it is still current</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query shows all the EPO alignments in this database: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM method_link_species_set WHERE method_link_id = 13;</pre> <a id="e_method_link_species_set" style="cursor:pointer;text-decoration:none" onclick="show_hide('method_link_species_set','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_method_link_species_set" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>method_link_species_set_id</th><th>method_link_id</th><th>species_set_id</th><th>name</th><th>source</th><th>url</th><th>first_release</th><th>last_release</th></tr>      <tr><td>647</td><td>13</td><td>35399</td><td>4 sauropsids EPO</td><td>ensembl</td><td></td><td>74</td><td></td></tr>      <tr class="bg2"><td>756</td><td>13</td><td>35886</td><td>8 primates EPO</td><td>ensembl</td><td></td><td>77</td><td></td></tr>      <tr><td>768</td><td>13</td><td>36100</td><td>5 teleost fish EPO</td><td>ensembl</td><td>mysql://ensro@compara4:3306/sf5_epo_5fish_79</td><td>80</td><td></td></tr>      <tr class="bg2"><td>780</td><td>13</td><td>36102</td><td>17 eutherian mammals EPO</td><td>ensembl</td><td>mysql://ensro@compara4:3306/mp14_epo_17mammals_80</td><td>80</td><td></td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link">method_link</a></li>
    <li><a href="#species_set">species_set</a></li>
  </ul>


  <div id="method_link_species_set_tag" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">method_link_species_set_tag</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_method_link_species_set_tag" style="cursor:pointer;text-decoration:none" onclick="show_hide('method_link_species_set_tag')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Contains serveral tag/value data associated with method_link_species_set entries</p>

  <div id="div_method_link_species_set_tag" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>tag</b></td><td>varchar(50)</td><td></td><td>Tag name</td><td>primary key</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>mediumtext</td><td></td><td>Tag value</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
  </ul>


  <div id="species_tree_node" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">species_tree_node</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_species_tree_node" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_tree_node')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the nodes of the species tree used in the gene gain/loss analysis</p>

  <div id="div_species_tree_node" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>node_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>parent_id</b></td><td>int(10)</td><td></td><td>Link to the parent node</td><td>key: parent_id</td></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>int(10)</td><td></td><td>Link to the root node</td><td>key: root_id</td></tr>
      <tr class="bg2"><td><b>left_index</b></td><td>int(10)</td><td>0</td><td>Internal index</td><td>key: root_id</td></tr>
      <tr class="bg1"><td><b>right_index</b></td><td>int(10)</td><td>0</td><td>Internal index</td><td></td></tr>
      <tr class="bg2"><td><b>distance_to_parent</b></td><td>double</td><td>'1'</td><td>Phylogenetic distance between this node and its parent</td><td></td></tr>
      <tr class="bg1"><td><b>taxon_id</b></td><td>int(10)</td><td></td><td>Link to NCBI taxon node</td><td></td></tr>
      <tr class="bg2"><td><b>genome_db_id</b></td><td>int(10)</td><td></td><td>Link to the genome_db</td><td></td></tr>
      <tr class="bg1"><td><b>node_name</b></td><td>VARCHAR(255)</td><td></td><td>A name that can be set to the taxon name or any other arbitrary name</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#species_tree_node_tag">species_tree_node_tag</a></li>
    <li><a href="#species_tree_root">species_tree_root</a></li>
    <li><a href="#CAFE_gene_family">CAFE_gene_family</a></li>
    <li><a href="#CAFE_species_gene">CAFE_species_gene</a></li>
  </ul>


  <div id="species_tree_root" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">species_tree_root</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_species_tree_root" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_tree_root')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores species trees used in compara. Each tree is made of species_tree_node's</p>

  <div id="div_species_tree_root" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table</td><td>unique: key</td></tr>
      <tr class="bg1"><td><b>label</b></td><td>VARCHAR(256)</td><td>'default'</td><td>Label to differentiate different trees with the same mlss_id</td><td>unique: key</td></tr>
      <tr class="bg2"><td><b>species_tree</b></td><td>mediumtext</td><td></td><td>Newick formatted version of the whole species_tree</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">Retrieve all the species trees stored in the database </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM species_tree_root</pre> <a id="e_species_tree_root" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_tree_root','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_species_tree_root" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>root_id</th><th>method_link_species_set_id</th><th>label</th><th>species_tree</th></tr>      <tr><td>647000</td><td>647</td><td>default</td><td>(((Gallus_gallus:0.0414,Meleagris_gallopavo:0.0414):0.1242,Taeniopygia_guttata:0.1715):0.3044,Anolis_carolinensis:0.4989):0;</td></tr>      <tr class="bg2"><td>664000</td><td>664</td><td>default</td><td>(((((Gallus_gallus:0.0414,Meleagris_gallopavo:0.0414):0.0414,Anas_platyrhynchos:0.0828):0.0828,(Taeniopygia_guttata:0.0858,Ficedula_albicollis:0.0858):0.0858):0.1992,Pelodiscus_sinensis:0.4892):0.1051,Anolis_carolinensis:0.4989):0;</td></tr>      <tr><td>756000</td><td>756</td><td>default</td><td>(((((Homo_sapiens:0.0059,Pan_troglodytes:0.0064):0.0019,Gorilla_gorilla:0.0086):0.0080,Pongo_abelii:0.017):0.0121,((Papio_anubis:0.0088,Macaca_mulatta:0.0088):0.0132,Chlorocebus_sabaeus:0.0132):0.0132):0.0147,Callithrix_jacchus:0.0687):0;</td></tr>      <tr class="bg2"><td>768000</td><td>768</td><td>default</td><td>((((Gasterosteus_aculeatus:0.14,Tetraodon_nigroviridis:0.1907):0.068,Oryzias_latipes:0.2212):0.0985,Danio_rerio:0.2706):0.107,Lepisosteus_oculatus:0.107):0;</td></tr>      <tr><td>780000</td><td>780</td><td>default</td><td>((((Mus_musculus:0.0765,Rattus_norvegicus:0.0834):0.2797,Oryctolagus_cuniculus:0.221):0.0095,(((((Homo_sapiens:0.0059,Pan_troglodytes:0.0064):0.0019,Gorilla_gorilla:0.0086):0.0080,Pongo_abelii:0.017):0.0121,((Papio_anubis:0.0088,Macaca_mulatta:0.0088):0.0132,Chlorocebus_sabaeus:0.0132):0.0132):0.0147,Callithrix_jacchus:0.0687):0.0795):0.0199,(((Felis_catus:0.0889,Canis_familiaris:0.0993):0.0505,Equus_caballus:0.0981):0.0067,((Bos_taurus:0.0261,Ovis_aries:0.0313):0.1104,Sus_scrofa:0.117):0.0318):0.0315):0;</td></tr>      <tr class="bg2"><td>781000</td><td>781</td><td>default</td><td>(((((((((Ailuropoda_melanoleuca:0.0629,Mustela_putorius_furo:0.0928):0.0233,Canis_familiaris:0.0837):0.0156,Felis_catus:0.0889):0.0505,Equus_caballus:0.0981):0.0053,(Pteropus_vampyrus:0.1122,Myotis_lucifugus:0.1269):0.0253):0.0014,((((Bos_taurus:0.0261,Ovis_aries:0.0313):0.0877,Tursiops_truncatus:0.0712):0.0127,Vicugna_pacos:0.1197):0.01,Sus_scrofa:0.117):0.0318):0.0128,(Erinaceus_europaeus:0.2392,Sorex_araneus:0.2565):0.0341):0.0187,((((((((((Homo_sapiens:0.0059,Pan_troglodytes:0.0064):0.0019,Gorilla_gorilla:0.0086):0.008,Pongo_abelii:0.017):0.0026,Nomascus_leucogenys:0.0191):0.0095,((Papio_anubis:0.0088,Macaca_mulatta:0.0088):0.0132,Chlorocebus_sabaeus:0.0132):0.0132):0.0147,Callithrix_jacchus:0.0687):0.0565,Tarsius_syrichta:0.1392):0.0083,(Microcebus_murinus:0.0849,Otolemur_garnettii:0.132):0.0359):0.0116,Tupaia_belangeri:0.1962):0.0031,(((((Mus_musculus:0.0765,Rattus_norvegicus:0.0834):0.2377,Dipodomys_ordii:0.236):0.0171,Ictidomys_tridecemlineatus:0.1728):0.0041,Cavia_porcellus:0.2403):0.0208,(Oryctolagus_cuniculus:0.1333,Ochotona_princeps:0.2054):0.0877):0.0095):0.0199):0.0117,(((Loxodonta_africana:0.0854,Procavia_capensis:0.162):0.0288,Echinops_telfairi:0.2362):0.0459,(Choloepus_hoffmanni:0.1086,Dasypus_novemcinctus:0.1249):0.0537):0.0117):0;</td></tr>      <tr><td>788000</td><td>788</td><td>default</td><td>(((((((Mus_musculus:0.0845,Rattus_norvegicus:0.0916):0.2567,Oryctolagus_cuniculus:0.2157):0.0153,((((Papio_anubis:0.01875,Macaca_mulatta:0.01875):0.009375,Chlorocebus_sabaeus:0.009375):0.009375,(((Pan_troglodytes:0.0067,Homo_sapiens:0.0067):0.0022,Gorilla_gorilla:0.0088):0.0097,Pongo_abelii:0.0183):0.0143):0.022,Callithrix_jacchus:0.0661):0.0891):0.0206,(((Felis_catus:0.0986,Canis_familiaris:0.1025):0.0498,Equus_caballus:0.1094):0.0107,((Bos_taurus:0.0618,Ovis_aries:0.0618):0.0869,Sus_scrofa:0.1073):0.0403):0.0329):0.2584,Monodelphis_domestica:0.3408):0.0717,Ornithorhynchus_anatinus:0.4566):0.1095,(((Gallus_gallus:0.0414,Meleagris_gallopavo:0.0414):0.1242,Taeniopygia_guttata:0.1715):0.3044,Anolis_carolinensis:0.4989):0.17):0;</td></tr>      <tr class="bg2"><td>793000</td><td>793</td><td>default</td><td>((((((((Xiphophorus_maculatus:0.10185,Poecilia_formosa:0.10185):0.10185,Oryzias_latipes:0.14):0.0483,Gasterosteus_aculeatus:0.1836):0.001,Oreochromis_niloticus:0.1728):0.0319,(Takifugu_rubripes:0.1096,Tetraodon_nigroviridis:0.1227):0.136):0.0411,Gadus_morhua:0.1965):0.0574,(Astyanax_mexicanus:0.1487,Danio_rerio:0.2053):0.0653):0.107,Lepisosteus_oculatus:0.107):0;</td></tr>      <tr><td>40102000</td><td>40102</td><td>default</td><td>(((6239,7227)1206794,(((((((((9315,9305,13616)9263,(43179,9544,10090,30538,9669,132908,9361,9940,9685,10020,9555,42254,9595,10116,9483,9478,9615,9986,9365,61853,9646,9371,9606,9823,30608,9913,9601,59463,37347,30611,9358,9785,60711,9813,9598,10141,9978,9739,9796)9347)32525,9258)40674,(9103,59729,13735,59894,9031,8839,28377)32561)32524,8364)32523,7897)8287,((8090,7955,48698,69293,31033,99883,7994,8128,8083,8049)186625,7918)41665)117571,7757)7742,(51511,7719)7718)7711)33213,4932)33154;</td></tr>      <tr class="bg2"><td>40102086</td><td>40102</td><td>full_species_tree</td><td>(((6239,7227)1206794,(((((((((9305,9315,13616)9263,((((((((((9598,9606,9595)207598,9601)9604,61853)314295,(9555,60711,9544)9528)9526,9483)314293,9478)376913,(30611,30608)376911)9443,((((10116,10090)39107,10020,43179)33553,10141)9989,(9986,9978)9975)314147,37347)314146,((59463,132908)9397,((9646,9615,9669)379584,9685)33554,(9365,42254)9362,((9913,9940)9895,9823,30538,9739)91561,9796)314145)1437010,(9361,9358)9348,(9813,9785,9371)311790)9347)32525,9258)40674,(((((9031,9103)9005,8839)1549675,(59894,59729)9126)8825,13735)1329799,28377)32561)32524,8364)32523,7897)8287,(((7955,7994)186626,(((((8083,48698)586240,8090)1489913,8128)1489908,((99883,31033)31031,69293)1489922)1489872,8049)123368)186625,7918)41665)117571,7757)7742,(7719,51511)7718)7711)33213,4932)33154;</td></tr>      <tr><td>40102210</td><td>40102</td><td>cafe</td><td>(((((homo.sapiens:2,pan.troglodytes:2)Homininae:6,callithrix.jacchus:8)Simiiformes:10,mus.musculus:18)Euarchontoglires.dup3:31,taeniopygia.guttata:49)Amniota:13,danio.rerio:62)Euteleostomi:0;</td></tr>      <tr class="bg2"><td>40101000</td><td>40101</td><td>default</td><td>(((6239,7227)1206794,((((((((((9361,9358)9348,(((((((((9598,9606,9595)207598,9601)9604,61853)314295,(60711,9555,9544)9528)9526,9483)314293,9478)376913,(30611,30608)376911)9443,((9978,9986)9975,(((10116,10090)39107,10020,43179)33553,10141)9989)314147,37347)314146,((42254,9365)9362,((9646,9615,9669)379584,9685)33554,(59463,132908)9397,((9913,9940)9895,9823,30538,9739)91561,9796)314145)1437010,(9371,9785,9813)311790)9347,(9315,9305,13616)9263)32525,9258)40674,((((59894,59729)9126,((9103,9031)9005,8839)1549675)8825,13735)1329799,28377)32561)32524,8364)32523,7897)8287,(((7955,7994)186626,((((31033,99883)31031,69293)1489922,(((8083,48698)586240,8090)1489913,8128)1489908)1489872,8049)123368)186625,7918)41665)117571,7757)7742,(51511,7719)7718)7711)33213,4932)33154;</td></tr>      <tr><td>40101124</td><td>40101</td><td>cafe</td><td>(((((((((((((((((((((((gorilla.gorilla:1,pan.troglodytes:1)Homininae.dup3:1,homo.sapiens:2)Homininae:1,pongo.abelii:3)Hominidae:1,nomascus.leucogenys:4)Hominoidea:3,((chlorocebus.sabaeus:1,macaca.mulatta:1)Cercopithecinae.dup3:1,papio.anubis:2)Cercopithecinae:5)Catarrhini:1,callithrix.jacchus:8)Simiiformes:1,tarsius.syrichta:9)Haplorrhini:2,(otolemur.garnettii:1,microcebus.murinus:1)Strepsirrhini:10)Primates:7,(((((rattus.norvegicus:1,mus.musculus:1)Murinae:1,ictidomys.tridecemlineatus:2)Sciurognathi.dup3:1,dipodomys.ordii:3)Sciurognathi:1,cavia.porcellus:4)Rodentia:2,(ochotona.princeps:1,oryctolagus.cuniculus:1)Lagomorpha:5)Glires:12)Euarchontoglires.dup3:1,tupaia.belangeri:19)Euarchontoglires:14,((((((bos.taurus:1,ovis.aries:1)Bovidae:1,tursiops.truncatus:2)Cetartiodactyla.dup3:1,vicugna.pacos:3)Cetartiodactyla.dup4:1,sus.scrofa:4)Cetartiodactyla:5,(((sorex.araneus:1,erinaceus.europaeus:1)Insectivora:1,equus.caballus:2)Laurasiatheria.dup3:2,(pteropus.vampyrus:1,myotis.lucifugus:1)Chiroptera:3)Laurasiatheria.dup4:5)Laurasiatheria.dup5:4,(((canis.familiaris:1,ailuropoda.melanoleuca:1)Caniformia.dup3:1,mustela.putorius.furo:2)Caniformia:1,felis.catus:3)Carnivora:10)Laurasiatheria:20)Boreoeutheria:2,(choloepus.hoffmanni:1,dasypus.novemcinctus:1)Xenarthra:34)Eutheria.dup3:3,((echinops.telfairi:1,procavia.capensis:1)Afrotheria.dup3:1,loxodonta.africana:2)Afrotheria:36)Eutheria:3,((macropus.eugenii:1,sarcophilus.harrisii:1)Marsupialia.dup3:1,monodelphis.domestica:2)Marsupialia:39)Theria:1,ornithorhynchus.anatinus:42)Mammalia:7,(((((gallus.gallus:1,meleagris.gallopavo:1)Phasianidae:1,anas.platyrhynchos:2)Galloanserae:2,(taeniopygia.guttata:1,ficedula.albicollis:1)Passeriformes:3)Neognathae:1,pelodiscus.sinensis:5)Archelosauria:1,anolis.carolinensis:6)Sauria:43)Amniota:1,xenopus.tropicalis:50)Tetrapoda:1,latimeria.chalumnae:51)Sarcopterygii:11,(((((((xiphophorus.maculatus:1,poecilia.formosa:1)Poeciliinae:1,oryzias.latipes:2)Atherinomorphae:1,oreochromis.niloticus:3)Ovalentaria:3,((tetraodon.nigroviridis:1,takifugu.rubripes:1)Tetraodontidae:1,gasterosteus.aculeatus:2)Eupercaria:4)Percomorphaceae:1,gadus.morhua:7)Acanthomorphata:2,(astyanax.mexicanus:1,danio.rerio:1)Otophysa:8)Clupeocephala:1,lepisosteus.oculatus:10)Neopterygii:52)Euteleostomi:1,petromyzon.marinus:63)Vertebrata:2,(ciona.savignyi:1,ciona.intestinalis:1)Ciona:64)Chordata:2,(drosophila.melanogaster:1,caenorhabditis.elegans:1)Ecdysozoa:66)Bilateria:1,saccharomyces.cerevisiae:68)root:0;</td></tr>      <tr class="bg2"><td>40102221</td><td>60003</td><td>Ensembl</td><td>((((9,8)7,((((((((((23,24)22,25)21,(20,19)18)17,26)16,27)15,(((((((45,44)43,((((50,51)49,52)48,53)47,54)46)42,(((((((((73,72)71,74)70,75)69,76)68,((66,65)64,67)63)62,77)61,78)60,(58,59)57)56,79)55)41,((((((102,103)101,104)100,105)99,106)98,(((((94,93)92,95)91,96)90,97)89,(88,87)86)85)84,(83,82)81)80)40,((38,39)37,((35,34)33,36)32)31)30,((110,109)108,111)107)29,112)28)14,113)13,114)12,(((118,119)117,((((((129,130)128,131)127,132)126,133)125,(123,124)122)121,134)120)116,135)115)11,136)10)6,(5,4)3)2,137)1;</td></tr>      <tr><td>40102358</td><td>60003</td><td>NCBI Taxonomy</td><td>((((((((((((9361,9358)9348,((((30608,30611)376911,((((((9598,9606,9595)207598,9601)9604,61853)314295,(9555,9544,60711)9528)9526,9483)314293,9478)376913)9443,((9986,9978)9975,(((10116,10090)39107,10020,43179)33553,10141)9989)314147,37347)314146,(((9646,9615,9669)379584,9685)33554,(59463,132908)9397,(42254,9365)9362,((9940,9913)9895,9823,30538,9739)91561,9796)314145)1437010,(9371,9813,9785)311790)9347,(9305,9315,13616)9263)32525,9258)40674,(((((9103,9031)9005,8839)1549675,(59894,59729)9126)8825,13735)1329799,28377)32561)32524,8364)32523,7897)8287,(((7955,7994)186626,(((((8083,48698)586240,8090)1489913,8128)1489908,((31033,99883)31031,69293)1489922)1489872,8049)123368)186625,7918)41665)117571,7757)7742,(7719,51511)7718)7711,(6239,7227)1206794)33213,4932)33154;</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#species_tree_node">species_tree_node</a></li>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
  </ul>


  <div id="species_tree_node_tag" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#3CB371;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">species_tree_node_tag</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_species_tree_node_tag" style="cursor:pointer;text-decoration:none" onclick="show_hide('species_tree_node_tag')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains tag/value data for species_tree_nodes</p>

  <div id="div_species_tree_node_tag" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>node_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>key: node_id_tag<br />key: tag_node_id<br />key: node_id</td></tr>
      <tr class="bg2"><td><b>tag</b></td><td>varchar(50)</td><td></td><td>Tag name for the tag/value pair</td><td>key: node_id_tag<br />key: tag_node_id<br />key: tag</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>mediumtext</td><td></td><td>Value for the tag/value pair</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#species_tree_node">species_tree_node</a></li>
    <li><a href="#species_tree_root">species_tree_root</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #FF8500">
  <h2 style="display:inline;color:#000">Genomic alignments tables</h2>
</div>
<p style="width:800px">These tables store information about genomic alignments in the Compara schema</p>
  <div id="synteny_region" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">synteny_region</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_synteny_region" style="cursor:pointer;text-decoration:none" onclick="show_hide('synteny_region')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Contains all the syntenic relationships found and the relative orientation of both syntenic regions.</p>

  <div id="div_synteny_region" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>synteny_region_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table</td><td>key: method_link_species_set_id</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query shows the 4 first syntenic regions between the Human and Opossum genomes by linking with the <a href="#method_link_species_set">method_link_species_set</a> table  </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT synteny_region.* FROM synteny_region JOIN method_link_species_set USING (method_link_species_set_id) JOIN species_set_header USING (species_set_id) WHERE species_set_header.name = "H.sap-M.dom" LIMIT 4;</pre> <a id="e_synteny_region" style="cursor:pointer;text-decoration:none" onclick="show_hide('synteny_region','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_synteny_region" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>synteny_region_id</th><th>method_link_species_set_id</th></tr>      <tr><td>48598</td><td>10096</td></tr>      <tr class="bg2"><td>48599</td><td>10096</td></tr>      <tr><td>48600</td><td>10096</td></tr>      <tr class="bg2"><td>48601</td><td>10096</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
  </ul>


  <div id="dnafrag" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">dnafrag</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_dnafrag" style="cursor:pointer;text-decoration:none" onclick="show_hide('dnafrag')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table defines the genomic sequences used in the comparative genomics analyisis. It is used by the <a href="#genomic_align_block">genomic_align_block</a> table to define aligned sequences. It is also used by the <a href="#dnafrag_region">dnafrag_region</a> table to define syntenic regions.<br />NOTE: Index <name> has genome_db_id in the first place because unless fetching all dnafrags or fetching by dnafrag_id, genome_db_id appears always in the WHERE clause. Unique key <name> is used to ensure that Bio::EnsEMBL::Compara::DBSQL::DnaFragAdaptor->fetch_by_GenomeDB_and_name will always fetch a single row. This can be used in the EnsEMBL Compara DB because we store top-level dnafrags only.</p>

  <div id="div_dnafrag" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>dnafrag_id</b></td><td>bigint</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>length</b></td><td>int(11)</td><td>0</td><td>The total length of the dnafrag</td><td></td></tr>
      <tr class="bg1"><td><b>name</b></td><td>varchar(40)</td><td>''</td><td>Name of the DNA sequence (e.g., the name of the chromosome)</td><td>unique: name</td></tr>
      <tr class="bg2"><td><b>genome_db_id</b></td><td>int(10)</td><td></td><td>External reference to genome_db_id in the <a href="#genome_db">genome_db</a> table</td><td>unique: name</td></tr>
      <tr class="bg1"><td><b>coord_system_name</b></td><td>varchar(40)</td><td>NULL</td><td>Refers to the coord system in which this dnafrag has been defined</td><td></td></tr>
      <tr class="bg2"><td><b>is_reference</b></td><td>tinyint(1)</td><td>1</td><td>Boolean, whether dnafrag is reference (1) or non-reference (0) eg haplotype</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">This query shows the chromosome 14 of the Human genome (genome_db.genome_db_id = 150 refers to Human genome in this example) which is 107349540 nucleotides long. </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT dnafrag.* FROM dnafrag LEFT JOIN genome_db USING (genome_db_id) WHERE dnafrag.name = "14" AND genome_db.name = "homo_sapiens";</pre> <a id="e_dnafrag" style="cursor:pointer;text-decoration:none" onclick="show_hide('dnafrag','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_dnafrag" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>dnafrag_id</th><th>length</th><th>name</th><th>genome_db_id</th><th>coord_system_name</th><th>is_reference</th></tr>      <tr><td>13955524</td><td>107043718</td><td>14</td><td>150</td><td>chromosome</td><td>1</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#genomic_align_block">genomic_align_block</a></li>
    <li><a href="#dnafrag_region">dnafrag_region</a></li>
    <li><a href="#genome_db">genome_db</a></li>
  </ul>


  <div id="dnafrag_region" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">dnafrag_region</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_dnafrag_region" style="cursor:pointer;text-decoration:none" onclick="show_hide('dnafrag_region')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the genomic regions corresponding to every synteny relationship found. There are two genomic regions for every synteny relationship.</p>

  <div id="div_dnafrag_region" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>synteny_region_id</b></td><td>int(10)</td><td>0</td><td>External reference to synteny_region_id in the <a href="#synteny_region">synteny_region</a> table</td><td>key: synteny<br />key: synteny_reversed</td></tr>
      <tr class="bg2"><td><b>dnafrag_id</b></td><td>bigint</td><td>0</td><td>External reference to dnafrag_id in the <a href="#dnafrag">dnafrag</a> table</td><td>key: synteny<br />key: synteny_reversed</td></tr>
      <tr class="bg1"><td><b>dnafrag_start</b></td><td>int(10)</td><td>0</td><td>Position of the first nucleotide from this dnafrag which is in synteny</td><td></td></tr>
      <tr class="bg2"><td><b>dnafrag_end</b></td><td>int(10)</td><td>0</td><td>Position of the last nucleotide from this dnafrag which is in synteny</td><td></td></tr>
      <tr class="bg1"><td><b>dnafrag_strand</b></td><td>tinyint(4)</td><td>0</td><td>Strand of this region</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">Return two top dnafrag regions </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM dnafrag_region ORDER BY synteny_region_id LIMIT 2;</pre> <a id="e_dnafrag_region1" style="cursor:pointer;text-decoration:none" onclick="show_hide('dnafrag_region1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_dnafrag_region1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>synteny_region_id</th><th>dnafrag_id</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strand</th></tr>      <tr><td>41285</td><td>13200496</td><td>21175813</td><td>21568912</td><td>-1</td></tr>      <tr class="bg2"><td>41285</td><td>13701542</td><td>1443</td><td>341228</td><td>1</td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">Same two dnafrag regions, but annotated </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT genome_db.name AS species_name, dnafrag.name, dnafrag_start, dnafrag_end, dnafrag_strand FROM dnafrag_region JOIN dnafrag USING (dnafrag_id) JOIN genome_db USING (genome_db_id) ORDER BY synteny_region_id LIMIT 2;</pre> <a id="e_dnafrag_region2" style="cursor:pointer;text-decoration:none" onclick="show_hide('dnafrag_region2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_dnafrag_region2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>species_name</th><th>dnafrag.name</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strand</th></tr>      <tr><td>bos_taurus</td><td>25</td><td>21175813</td><td>21568912</td><td>-1</td></tr>      <tr class="bg2"><td>sus_scrofa</td><td>10</td><td>1443</td><td>341228</td><td>1</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#synteny_region">synteny_region</a></li>
    <li><a href="#dnafrag">dnafrag</a></li>
  </ul>


  <div id="genomic_align_block" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">genomic_align_block</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_genomic_align_block" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align_block')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table is the key table for the genomic alignments. The software used to align the genomic blocks is refered as an external key to the <a href="#method_link">method_link</a> table. Nevertheless, actual aligned sequences are defined in the <a href="#genomic_align">genomic_align</a> table.<br />Tree alignments (EPO alignments) are best accessed through the <a href="#genomic_align_tree">genomic_align_tree</a> table although the alignments are also indexed in this table. This allows the user to also access the tree alignments as normal multiple alignments.<br />NOTE: All queries in the API uses the primary key as rows are always fetched using the genomic_align_block_id. The key 'method_link_species_set_id' is used by MART when fetching all the genomic_align_blocks corresponding to a given method_link_species_set_id</p>

  <div id="div_genomic_align_block" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>genomic_align_block_id</b></td><td>bigint</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td>0</td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table</td><td>key: method_link_species_set_id</td></tr>
      <tr class="bg1"><td><b>score</b></td><td>double</td><td></td><td>Score returned by the homology search program</td><td></td></tr>
      <tr class="bg2"><td><b>perc_id</b></td><td>tinyint(3)</td><td>NULL</td><td>Used for pairwise comparison. Defines the percentage of identity between both sequences</td><td></td></tr>
      <tr class="bg1"><td><b>length</b></td><td>int(10)</td><td></td><td>Total length of the alignment</td><td></td></tr>
      <tr class="bg2"><td><b>group_id</b></td><td>bigint</td><td>NULL</td><td>Used to group alignments</td><td></td></tr>
      <tr class="bg1"><td><b>level_id</b></td><td>tinyint(2)</td><td>0</td><td>Level of orthologous layer. 1 corresponds to the principal layer of orthologous sequences found (the largest), 2 and over are additional layers. Use for building the syntenies (based on level_id = 1 only). Note that level_ids are not computed on whole chromosomes but rather on chunks. This means that level_ids can be inconsistent within an alignment-net.</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query refers to the LastZ alignment between medaka and zebrafish: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT genomic_align_block.* FROM genomic_align_block JOIN method_link_species_set USING (method_link_species_set_id) JOIN species_set_header USING (species_set_id) WHERE method_link_id = 16 AND species_set_header.name = "O.lat-D.rer" ORDER BY genomic_align_block_id LIMIT 4;</pre> <a id="e_genomic_align_block" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align_block','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_genomic_align_block" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>genomic_align_block_id</th><th>method_link_species_set_id</th><th>score</th><th>perc_id</th><th>length</th><th>group_id</th><th>level_id</th></tr>      <tr><td>7720000000001</td><td>772</td><td>3028</td><td></td><td>203</td><td>7720032481357</td><td>2</td></tr>      <tr class="bg2"><td>7720000000002</td><td>772</td><td>4873</td><td></td><td>143</td><td>7720032480284</td><td>1</td></tr>      <tr><td>7720000000003</td><td>772</td><td>5577</td><td></td><td>188</td><td>7720032496522</td><td>1</td></tr>      <tr class="bg2"><td>7720000000004</td><td>772</td><td>42115</td><td></td><td>159</td><td>7720032496279</td><td>1</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
    <li><a href="#genomic_align_tree">genomic_align_tree</a></li>
    <li><a href="#method_link">method_link</a></li>
    <li><a href="#genomic_align">genomic_align</a></li>
  </ul>


  <div id="genomic_align_tree" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">genomic_align_tree</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_genomic_align_tree" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align_tree')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table is used to index tree alignments, e.g. EPO alignments. These alignments include inferred ancestral sequences. The tree required to index these sequences is stored in this table. This table stores the structure of the tree. Each node links to an entry in the <a href="#genomic_align_group">genomic_align_group</a> table, which links to one or several entries in the <a href="#genomic_align">genomic_align</a> table.<br />NOTE: Left_index and right_index are used to speed up fetching trees from the database. Any given node has its left_index larger than the left_index of its parent node and its right index smaller than the right_index of its parent node. In other words, all descendent nodes of a given node can be obtained by fetching all the node with a left_index (or right_index or both) between the left_index and the right_index of that node.</p>

  <div id="div_genomic_align_tree" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>node_id</b></td><td>bigint(20)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>parent_id</b></td><td>bigint(20)</td><td>0</td><td>Link to the parent node</td><td>key: parent_id</td></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>bigint(20)</td><td>0</td><td>Link to root node</td><td>key: root_id<br />key: left_index</td></tr>
      <tr class="bg2"><td><b>left_index</b></td><td>int(10)</td><td>0</td><td>Internal index. See above</td><td>key: left_index</td></tr>
      <tr class="bg1"><td><b>right_index</b></td><td>int(10)</td><td>0</td><td>Internal index. See above</td><td></td></tr>
      <tr class="bg2"><td><b>left_node_id</b></td><td>bigint(10)</td><td>0</td><td>Link to the node on the left side of this node</td><td></td></tr>
      <tr class="bg1"><td><b>right_node_id</b></td><td>bigint(10)</td><td>0</td><td>Link to the node on the right side of this node</td><td></td></tr>
      <tr class="bg2"><td><b>distance_to_parent</b></td><td>double</td><td>1</td><td>Phylogenetic distance between this node and its parent</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">The following query corresponds to the root of a tree, because parent_id = 0 and root_id = node_id </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM genomic_align_tree WHERE node_id = root_id LIMIT 1;</pre> <a id="e_genomic_align_tree1" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align_tree1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_genomic_align_tree1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>node_id</th><th>parent_id</th><th>root_id</th><th>left_index</th><th>right_index</th><th>left_node_id</th><th>right_node_id</th><th>distance_to_parent</th></tr>      <tr><td>6470000000001</td><td>0</td><td>6470000000001</td><td>600000001</td><td>600000006</td><td>0</td><td>0</td><td>0.3044</td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">Grab the first two trees that have exactly 5 nodes </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM genomic_align_tree WHERE root_id IN (SELECT root_id FROM genomic_align_tree GROUP BY root_id HAVING COUNT(*) = 5) LIMIT 10;</pre> <a id="e_genomic_align_tree2" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align_tree2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_genomic_align_tree2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>*FROMgenomic_align_treeWHEREroot_idIN(SELECTroot_id</th></tr>      <tr><td>6470000000004</td><td>0</td><td>6470000000004</td><td>600000001</td><td>600000010</td><td>0</td><td>6470000036708</td><td>0.3044</td></tr>      <tr class="bg2"><td>6470000000005</td><td>6470000000004</td><td>6470000000004</td><td>600000008</td><td>600000009</td><td>6470000057398</td><td>6470000036709</td><td>0.1715</td></tr>      <tr><td>6470000000006</td><td>6470000000004</td><td>6470000000004</td><td>600000002</td><td>600000007</td><td>0</td><td>6470000036710</td><td>0.1242</td></tr>      <tr class="bg2"><td>6470000000007</td><td>6470000000006</td><td>6470000000004</td><td>600000003</td><td>600000004</td><td>6470000013437</td><td>6470000036711</td><td>0.0414</td></tr>      <tr><td>6470000000008</td><td>6470000000006</td><td>6470000000004</td><td>600000005</td><td>600000006</td><td>6470000057397</td><td>6470000036712</td><td>0.0414</td></tr>      <tr class="bg2"><td>6470000000009</td><td>0</td><td>6470000000009</td><td>600000001</td><td>600000010</td><td>0</td><td>0</td><td>0.3044</td></tr>      <tr><td>6470000000010</td><td>6470000000009</td><td>6470000000009</td><td>600000008</td><td>600000009</td><td>0</td><td>6470000020536</td><td>0.1715</td></tr>      <tr class="bg2"><td>6470000000011</td><td>6470000000009</td><td>6470000000009</td><td>600000002</td><td>600000007</td><td>0</td><td>6470000033189</td><td>0.1242</td></tr>      <tr><td>6470000000012</td><td>6470000000011</td><td>6470000000009</td><td>600000003</td><td>600000004</td><td>6470000027038</td><td>6470000033191</td><td>0.0414</td></tr>      <tr class="bg2"><td>6470000000013</td><td>6470000000011</td><td>6470000000009</td><td>600000005</td><td>600000006</td><td>6470000015126</td><td>6470000033190</td><td>0.0414</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#genomic_align_group">genomic_align_group</a></li>
    <li><a href="#genomic_align">genomic_align</a></li>
  </ul>


  <div id="genomic_align" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">genomic_align</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_genomic_align" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the coordinates and all the information needed to rebuild genomic alignments. Every entry corresponds to one of the aligned sequences. It also contains an external key to the <a href="#method_link_species_set">method_link_species_set</a> which refers to the software and set of species used for getting the corresponding alignment. The aligned sequence is defined by an external reference to the <a href="#dnafrag">dnafrag</a> table, the starting and ending position within this dnafrag, the strand and a cigar_line.<br /> The original aligned sequence is not stored but it can be retrieved using the <b>cigar_line</b> field and the original sequence. The cigar line defines the sequence of matches/mismatches and deletions (or gaps). For example, this cigar line <b>2MD3M2D2M</b> will mean that the alignment contains 2 matches/mismatches, 1 deletion (number 1 is omitted in order to save some space), 3 matches/mismatches, 2 deletions and 2 matches/mismatches. If the original sequence is:<br /> <ul><li>Original sequence: AACGCTT</li></ul> The aligned sequence will be:<br /> <table> <caption>cigar line: 2MD3M2D2M</caption> <thead><tr> <th>M</th> <th>M</th> <th>D</th> <th>M</th> <th>M</th> <th>M</th> <th>D</th> <th>D</th> <th>M</th> <th>M</th> </tr> </thead> <tbody> <tr> <td>A</td> <td>A</td> <td>-</td> <td>C</td> <td>G</td> <td>C</td> <td>-</td> <td>-</td> <td>T</td> <td>T</td> </tr> </tbody> </table></p>

  <div id="div_genomic_align" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>genomic_align_id</b></td><td>bigint</td><td></td><td>Unique internal ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>genomic_align_block_id</b></td><td>bigint</td><td></td><td>External reference to genomic_align_block_id in the <a href="#genomic_align_block">genomic_align_block</a> table</td><td>key: genomic_align_block_id</td></tr>
      <tr class="bg1"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td>0</td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table. This information is redundant because it also appears in the <a href="#genomic_align_block">genomic_align_block</a> table but it is used to speed up the queries</td><td>key: method_link_species_set_id<br />key: dnafrag</td></tr>
      <tr class="bg2"><td><b>dnafrag_id</b></td><td>bigint</td><td>0</td><td>External reference to dnafrag_id in the <a href="#dnafrag">dnafrag</a> table</td><td>key: dnafrag</td></tr>
      <tr class="bg1"><td><b>dnafrag_start</b></td><td>int(10)</td><td>0</td><td>Starting position within the dnafrag defined by dnafrag_id</td><td>key: dnafrag</td></tr>
      <tr class="bg2"><td><b>dnafrag_end</b></td><td>int(10)</td><td>0</td><td>Ending position within the dnafrag defined by dnafrag_id</td><td>key: dnafrag</td></tr>
      <tr class="bg1"><td><b>dnafrag_strand</b></td><td>tinyint(4)</td><td>0</td><td>Strand in the dnafrag defined by dnafrag_id</td><td></td></tr>
      <tr class="bg2"><td><b>cigar_line</b></td><td>mediumtext</td><td></td><td>Internal description of the aligned sequence</td><td></td></tr>
      <tr class="bg1"><td><b>visible</b></td><td>tinyint(2)</td><td>1</td><td>Used in self alignments to ensure only one Bio::EnsEMBL::Compara::GenomicAlignBlock is visible when you have more than 1 block covering the same region</td><td></td></tr>
      <tr class="bg2"><td><b>node_id</b></td><td>bigint(20)</td><td>NULL</td><td>External reference to node_id in the <a href="#genomic_align_tree">genomic_align_tree</a> table</td><td>key: node_id</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">The following query corresponds to the 4x2 sequences included in the alignment described above (see <a href="#genomic_align_block">genomic_align_block</a> table description). </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT genomic_align.* FROM genomic_align_block JOIN method_link_species_set USING (method_link_species_set_id) JOIN species_set_header USING (species_set_id) JOIN genomic_align USING (genomic_align_block_id) WHERE method_link_id = 16 AND species_set_header.name = "O.lat-D.rer" ORDER BY genomic_align_block_id LIMIT 8;</pre> <a id="e_genomic_align1" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_genomic_align1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>genomic_align_id</th><th>genomic_align_block_id</th><th>method_link_species_set_id</th><th>dnafrag_id</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strand</th><th>cigar_line</th><th>visible</th><th>node_id</th></tr>      <tr><td>7720000000001</td><td>7720000000001</td><td>772</td><td>14025335</td><td>72669</td><td>72858</td><td>1</td><td>60M12D31MD99M</td><td>1</td><td></td></tr>      <tr class="bg2"><td>7720000000002</td><td>7720000000001</td><td>772</td><td>2039154</td><td>19212929</td><td>19213130</td><td>1</td><td>157MD45M</td><td>1</td><td></td></tr>      <tr><td>7720000000003</td><td>7720000000002</td><td>772</td><td>14025335</td><td>5621</td><td>5762</td><td>1</td><td>88MD54M</td><td>1</td><td></td></tr>      <tr class="bg2"><td>7720000000004</td><td>7720000000002</td><td>772</td><td>2039159</td><td>18011823</td><td>18011964</td><td>-1</td><td>78MD64M</td><td>1</td><td></td></tr>      <tr><td>7720000000005</td><td>7720000000003</td><td>772</td><td>14025335</td><td>49222</td><td>49409</td><td>1</td><td>188M</td><td>1</td><td></td></tr>      <tr class="bg2"><td>7720000000006</td><td>7720000000003</td><td>772</td><td>2039378</td><td>829</td><td>1010</td><td>1</td><td>60MD14M3D53M2D55M</td><td>1</td><td></td></tr>      <tr><td>7720000000007</td><td>7720000000004</td><td>772</td><td>14025335</td><td>111693</td><td>111820</td><td>1</td><td>37M29D25M2D66M</td><td>1</td><td></td></tr>      <tr class="bg2"><td>7720000000008</td><td>7720000000004</td><td>772</td><td>2039174</td><td>137734</td><td>137892</td><td>-1</td><td>159M</td><td>1</td><td></td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">Here is a better way to get this by joining the <a href="#dnafrag">dnafrag</a> and <a href="#genome_db">genome_db</a> tables: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT genome_db.name, dnafrag.name, dnafrag_start, dnafrag_end, dnafrag_strand str, cigar_line FROM genomic_align_block JOIN method_link_species_set USING (method_link_species_set_id) JOIN species_set_header USING (species_set_id) JOIN genomic_align USING (genomic_align_block_id) JOIN dnafrag USING (dnafrag_id) JOIN genome_db USING (genome_db_id) WHERE method_link_id = 16 AND species_set_header.name = "O.lat-D.rer" ORDER BY genomic_align_block_id LIMIT 8;</pre> <a id="e_genomic_align2" style="cursor:pointer;text-decoration:none" onclick="show_hide('genomic_align2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_genomic_align2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>genome_db.name</th><th>dnafrag.name</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strandstr</th><th>cigar_line</th></tr>      <tr><td>danio_rerio</td><td>KN149683.1</td><td>72669</td><td>72858</td><td>1</td><td>60M12D31MD99M</td></tr>      <tr class="bg2"><td>oryzias_latipes</td><td>4</td><td>19212929</td><td>19213130</td><td>1</td><td>157MD45M</td></tr>      <tr><td>danio_rerio</td><td>KN149683.1</td><td>5621</td><td>5762</td><td>1</td><td>88MD54M</td></tr>      <tr class="bg2"><td>oryzias_latipes</td><td>18</td><td>18011823</td><td>18011964</td><td>-1</td><td>78MD64M</td></tr>      <tr><td>danio_rerio</td><td>KN149683.1</td><td>49222</td><td>49409</td><td>1</td><td>188M</td></tr>      <tr class="bg2"><td>oryzias_latipes</td><td>scaffold831</td><td>829</td><td>1010</td><td>1</td><td>60MD14M3D53M2D55M</td></tr>      <tr><td>danio_rerio</td><td>KN149683.1</td><td>111693</td><td>111820</td><td>1</td><td>37M29D25M2D66M</td></tr>      <tr class="bg2"><td>oryzias_latipes</td><td>ultracontig231</td><td>137734</td><td>137892</td><td>-1</td><td>159M</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#genomic_align_block">genomic_align_block</a></li>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
    <li><a href="#dnafrag">dnafrag</a></li>
    <li><a href="#genomic_align_tree">genomic_align_tree</a></li>
    <li><a href="#genome_db">genome_db</a></li>
  </ul>


  <div id="conservation_score" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">conservation_score</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_conservation_score" style="cursor:pointer;text-decoration:none" onclick="show_hide('conservation_score')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains conservation scores calculated from the whole-genome multiple alignments stored in the <a href="#genomic_align_block">genomic_align_block</a> table. Several scores are stored per row. expected_score and diff_score are binary columns and you need to use the Perl API to access these data.</p>

  <div id="div_conservation_score" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>genomic_align_block_id</b></td><td>bigint</td><td></td><td>External reference to genomic_align_block_id in the <a href="#genomic_align_block">genomic_align_block</a> table</td><td>key: genomic_align_block_id, window_size</td></tr>
      <tr class="bg2"><td><b>window_size</b></td><td>smallint</td><td></td><td>The scores are stored at different resolution levels. This column defines the window size used to calculate the average score</td><td>key: genomic_align_block_id, window_size</td></tr>
      <tr class="bg1"><td><b>position</b></td><td>int</td><td></td><td>Position of the first score (in alignment coordinates)</td><td></td></tr>
      <tr class="bg2"><td><b>expected_score</b></td><td>blob</td><td></td><td>Expected score. The observed score can be determined using the diff_score and the expected_score</td><td></td></tr>
      <tr class="bg1"><td><b>diff_score</b></td><td>blob</td><td></td><td>The difference between the expected and observed variation, i.e. the conservation score</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#genomic_align_block">genomic_align_block</a></li>
  </ul>


  <div id="constrained_element" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#FF8500;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">constrained_element</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_constrained_element" style="cursor:pointer;text-decoration:none" onclick="show_hide('constrained_element')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains constrained elements calculated from the whole-genome multiple alignments stored in the <a href="#genomic_align_block">genomic_align_block</a> table</p>

  <div id="div_constrained_element" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>constrained_element_id</b></td><td>bigint(20)</td><td></td><td>Internal (but unique) ID</td><td>key: constrained_element_id_idx</td></tr>
      <tr class="bg2"><td><b>dnafrag_id</b></td><td>bigint</td><td></td><td>External reference to dnafrag_id in the <a href="#dnafrag">dnafrag</a> table</td><td>key: mlssid_dfid_dfstart_dfend_idx<br />key: mlssid_dfid_idx</td></tr>
      <tr class="bg1"><td><b>dnafrag_start</b></td><td>int(12)</td><td></td><td>Start of the constrained element</td><td>key: mlssid_dfid_dfstart_dfend_idx</td></tr>
      <tr class="bg2"><td><b>dnafrag_end</b></td><td>int(12)</td><td></td><td>End of the constrained element</td><td>key: mlssid_dfid_dfstart_dfend_idx</td></tr>
      <tr class="bg1"><td><b>dnafrag_strand</b></td><td>int(2)</td><td></td><td>Strand of the constrained element</td><td></td></tr>
      <tr class="bg2"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table</td><td>key: mlssid_idx<br />key: mlssid_dfid_dfstart_dfend_idx<br />key: mlssid_dfid_idx</td></tr>
      <tr class="bg1"><td><b>p_value</b></td><td>double</td><td></td><td>p-value derived from Gerp</td><td></td></tr>
      <tr class="bg2"><td><b>score</b></td><td>double</td><td>0</td><td>Score derived from Gerp</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">Example entry for a constrained_element: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM constrained_element ORDER BY constrained_element_id LIMIT 1;</pre> <a id="e_constrained_element1" style="cursor:pointer;text-decoration:none" onclick="show_hide('constrained_element1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_constrained_element1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>constrained_element_id</th><th>dnafrag_id</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strand</th><th>method_link_species_set_id</th><th>p_value</th><th>score</th></tr>      <tr><td>6650000000001</td><td>13715906</td><td>2687447</td><td>2687748</td><td>-1</td><td>665</td><td>7.10339e-43</td><td>118.1</td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">There are 2 other elements in the same constrained_element: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM constrained_element JOIN (SELECT MIN(constrained_element_id) AS constrained_element_id FROM constrained_element) t USING (constrained_element_id);</pre> <a id="e_constrained_element2" style="cursor:pointer;text-decoration:none" onclick="show_hide('constrained_element2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_constrained_element2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>constrained_element_id</th></tr>      <tr><td>6650000000001</td><td>13715906</td><td>2687447</td><td>2687748</td><td>-1</td><td>665</td><td>7.10339e-43</td><td>118.1</td></tr>      <tr class="bg2"><td>6650000000001</td><td>12160494</td><td>46741244</td><td>46741528</td><td>-1</td><td>665</td><td>7.10339e-43</td><td>118.1</td></tr>      <tr><td>6650000000001</td><td>13862678</td><td>1522791</td><td>1523075</td><td>-1</td><td>665</td><td>7.10339e-43</td><td>118.1</td></tr>      <tr class="bg2"><td>6650000000001</td><td>13785054</td><td>1576786</td><td>1577070</td><td>-1</td><td>665</td><td>7.10339e-43</td><td>118.1</td></tr>      <tr><td>6650000000001</td><td>13777393</td><td>44982316</td><td>44982600</td><td>-1</td><td>665</td><td>7.10339e-43</td><td>118.1</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#dnafrag">dnafrag</a></li>
    <li><a href="#genomic_align_block">genomic_align_block</a></li>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
  </ul>


<br /><br />
<div style="background-color:#F2F2F2;padding:5px;margin:5px 0px;border-top:2px solid #1E90FF">
  <h2 style="display:inline;color:#000">Gene trees and homologies tables</h2>
</div>
<p style="width:800px">These tables store information about gene alignments, trees and homologies</p>
  <div id="sequence" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">sequence</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_sequence" style="cursor:pointer;text-decoration:none" onclick="show_hide('sequence')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the sequences of the seq_member entries</p>

  <div id="div_sequence" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>sequence_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>length</b></td><td>int(10)</td><td></td><td>Length of the sequence</td><td></td></tr>
      <tr class="bg1"><td><b>sequence</b></td><td>longtext</td><td></td><td>The actual sequence</td><td>key: sequence</td></tr>
    </table>
  </div>

  <div id="gene_member" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_member</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_member')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table links sequences to the EnsEMBL core DB or to external DBs.</p>

  <div id="div_gene_member" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>gene_member_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>stable_id</b></td><td>varchar(128)</td><td></td><td>EnsEMBL stable ID</td><td>unique: stable_id</td></tr>
      <tr class="bg1"><td><b>version</b></td><td>int(10)</td><td>0</td><td>Version of the stable ID (see EnsEMBL core DB)</td><td></td></tr>
      <tr class="bg2"><td><b>source_name</b></td><td>ENUM('ENSEMBLGENE', 'EXTERNALGENE')</td><td></td><td>The source of the member</td><td>key: source_name</td></tr>
      <tr class="bg1"><td><b>taxon_id</b></td><td>int(10)</td><td></td><td>External reference to taxon_id in the <a href="#ncbi_taxa_node">ncbi_taxa_node</a> table</td><td></td></tr>
      <tr class="bg2"><td><b>genome_db_id</b></td><td>int(10)</td><td></td><td>External reference to genome_db_id in the <a href="#genome_db">genome_db</a> table</td><td></td></tr>
      <tr class="bg1"><td><b>canonical_member_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table to allow linkage from a gene to its canonical peptide</td><td>key: canonical_member_id</td></tr>
      <tr class="bg2"><td><b>description</b></td><td>text</td><td>NULL</td><td>The description of the gene/protein as described in the core database or from the Uniprot entry</td><td></td></tr>
      <tr class="bg1"><td><b>dnafrag_id</b></td><td>bigint</td><td></td><td>External reference to dnafrag_id in the <a href="#dnafrag">dnafrag</a> table. It shows the dnafrag the member is on.</td><td>key: dnafrag_id_start<br />key: dnafrag_id_end</td></tr>
      <tr class="bg2"><td><b>dnafrag_start</b></td><td>int(10)</td><td></td><td>Starting position within the dnafrag defined by dnafrag_id</td><td>key: dnafrag_id_start</td></tr>
      <tr class="bg1"><td><b>dnafrag_end</b></td><td>int(10)</td><td></td><td>Ending position within the dnafrag defined by dnafrag_id</td><td>key: dnafrag_id_end</td></tr>
      <tr class="bg2"><td><b>dnafrag_strand</b></td><td>tinyint(4)</td><td></td><td>Strand in the dnafrag defined by dnafrag_id</td><td></td></tr>
      <tr class="bg1"><td><b>display_label</b></td><td>varchar(128)</td><td>NULL</td><td>Display name (imported from the core database)</td><td></td></tr>
      <tr class="bg2"><td><b>families</b></td><td>tinyint(1)</td><td>0</td><td>The number of families associated with this member</td><td></td></tr>
      <tr class="bg1"><td><b>gene_trees</b></td><td>tinyint(1)</td><td>0</td><td>If this member is part of a gene tree</td><td></td></tr>
      <tr class="bg2"><td><b>gene_gain_loss_trees</b></td><td>tinyint(1)</td><td>0</td><td>If this member is part of a gene gain/loss tree</td><td></td></tr>
      <tr class="bg1"><td><b>orthologues</b></td><td>int(10)</td><td>0</td><td>The number of orthologues for this member</td><td></td></tr>
      <tr class="bg2"><td><b>paralogues</b></td><td>int(10)</td><td>0</td><td>The number of paralogues for this member</td><td></td></tr>
      <tr class="bg1"><td><b>homoeologues</b></td><td>int(10)</td><td>0</td><td>The number of homoeologues for this member</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query refers to the human (ncbi_taxa_node.taxon_id = 9606 or genome_db_id = 150) gene ENSG00000173213 </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM gene_member WHERE stable_id = "ENSG00000173213";</pre> <a id="e_gene_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_member','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_gene_member" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>gene_member_id</th><th>stable_id</th><th>version</th><th>source_name</th><th>taxon_id</th><th>genome_db_id</th><th>canonical_member_id</th><th>description</th><th>dnafrag_id</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strand</th><th>display_label</th><th>families</th><th>gene_trees</th><th>gene_gain_loss_trees</th><th>orthologues</th><th>paralogues</th><th>homoeologues</th></tr>      <tr><td>9268897</td><td>ENSG00000173213</td><td>0</td><td>ENSEMBLGENE</td><td>9606</td><td>150</td><td>9645286</td><td>tubulin beta 8 pseudogene 12 [Source:HGNC Symbol;Acc:HGNC:24983]</td><td>13955525</td><td>47221</td><td>73545</td><td>-1</td><td>TUBB8P12</td><td>1</td><td>1</td><td>1</td><td>64</td><td>12</td><td>0</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#sequence">sequence</a></li>
    <li><a href="#ncbi_taxa_node">ncbi_taxa_node</a></li>
    <li><a href="#genome_db">genome_db</a></li>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#dnafrag">dnafrag</a></li>
  </ul>


  <div id="seq_member" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">seq_member</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_seq_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('seq_member')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table links sequences to the EnsEMBL core DB or to external DBs.</p>

  <div id="div_seq_member" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key<br />key: seq_member_gene_member_id_end</td></tr>
      <tr class="bg2"><td><b>stable_id</b></td><td>varchar(128)</td><td></td><td>EnsEMBL stable ID or external ID (for Uniprot/SWISSPROT and Uniprot/SPTREMBL)</td><td>unique: stable_id</td></tr>
      <tr class="bg1"><td><b>version</b></td><td>int(10)</td><td>0</td><td>Version of the stable ID (see EnsEMBL core DB)</td><td></td></tr>
      <tr class="bg2"><td><b>source_name</b></td><td>ENUM('ENSEMBLPEP','ENSEMBLTRANS','Uniprot/SPTREMBL','Uniprot/SWISSPROT','EXTERNALPEP','EXTERNALTRANS','EXTERNALCDS')</td><td></td><td>The source of the member</td><td>key: source_name</td></tr>
      <tr class="bg1"><td><b>taxon_id</b></td><td>int(10)</td><td></td><td>External reference to taxon_id in the <a href="#ncbi_taxa_node">ncbi_taxa_node</a> table</td><td></td></tr>
      <tr class="bg2"><td><b>genome_db_id</b></td><td>int(10)</td><td></td><td>External reference to genome_db_id in the <a href="#genome_db">genome_db</a> table</td><td></td></tr>
      <tr class="bg1"><td><b>sequence_id</b></td><td>int(10)</td><td></td><td>External reference to sequence_id in the <a href="#sequence">sequence</a> table. May be 0 when the sequence is not available in the <a href="#sequence">sequence</a> table, e.g. for a gene instance</td><td>key: sequence_id</td></tr>
      <tr class="bg2"><td><b>gene_member_id</b></td><td>int(10)</td><td></td><td>External reference to gene_member_id in the <a href="#gene_member">gene_member</a> table to allow linkage from peptides and transcripts to genes</td><td>key: gene_member_id<br />key: seq_member_gene_member_id_end</td></tr>
      <tr class="bg1"><td><b>description</b></td><td>text</td><td>NULL</td><td>The description of the gene/protein as described in the core database or from the Uniprot entry</td><td></td></tr>
      <tr class="bg2"><td><b>dnafrag_id</b></td><td>bigint</td><td></td><td>External reference to dnafrag_id in the <a href="#dnafrag">dnafrag</a> table. It shows the dnafrag the member is on.</td><td>key: dnafrag_id_start<br />key: dnafrag_id_end</td></tr>
      <tr class="bg1"><td><b>dnafrag_start</b></td><td>int(10)</td><td></td><td>Starting position within the dnafrag defined by dnafrag_id</td><td>key: dnafrag_id_start</td></tr>
      <tr class="bg2"><td><b>dnafrag_end</b></td><td>int(10)</td><td></td><td>Ending position within the dnafrag defined by dnafrag_id</td><td>key: dnafrag_id_end</td></tr>
      <tr class="bg1"><td><b>dnafrag_strand</b></td><td>tinyint(4)</td><td></td><td>Strand in the dnafrag defined by dnafrag_id</td><td></td></tr>
      <tr class="bg2"><td><b>display_label</b></td><td>varchar(128)</td><td>NULL</td><td>Display name (imported from the core database)</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query refers to the human (ncbi_taxa_node.taxon_id = 9606 or genome_db_id = 150) peptide ENSP00000309431 </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM seq_member WHERE stable_id = "ENSP00000309431";</pre> <a id="e_seq_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('seq_member','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_seq_member" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>seq_member_id</th><th>stable_id</th><th>version</th><th>source_name</th><th>taxon_id</th><th>genome_db_id</th><th>sequence_id</th><th>gene_member_id</th><th>description</th><th>dnafrag_id</th><th>dnafrag_start</th><th>dnafrag_end</th><th>dnafrag_strand</th><th>display_label</th></tr>      <tr><td>9645286</td><td>ENSP00000309431</td><td>0</td><td>ENSEMBLPEP</td><td>9606</td><td>150</td><td>4654669</td><td>9268897</td><td></td><td>13955525</td><td>47390</td><td>49850</td><td>-1</td><td>TUBB8P12-201</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#sequence">sequence</a></li>
    <li><a href="#ncbi_taxa_node">ncbi_taxa_node</a></li>
    <li><a href="#genome_db">genome_db</a></li>
    <li><a href="#gene_member">gene_member</a></li>
    <li><a href="#dnafrag">dnafrag</a></li>
  </ul>


  <div id="external_db" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">external_db</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_external_db" style="cursor:pointer;text-decoration:none" onclick="show_hide('external_db')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores data about the external databases in which the objects described in the <a href="#member_xref">member_xref</a> table are stored.</p>

  <div id="div_external_db" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>external_db_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>db_name</b></td><td>varchar(100)</td><td></td><td>External database name</td><td>unique key: db_name_db_release_idx</td></tr>
      <tr class="bg1"><td><b>db_release</b></td><td>varchar(255)</td><td>NULL</td><td>Release information about the external database</td><td>unique key: db_name_db_release_idx</td></tr>
      <tr class="bg2"><td><b>status</b></td><td>enum('KNOWNXREF','KNOWN','XREF','PRED','ORTH','PSEUDO')</td><td></td><td>Status, e.g. 'KNOWNXREF','KNOWN','XREF','PRED','ORTH','PSEUDO'.</td><td></td></tr>
      <tr class="bg1"><td><b>priority</b></td><td>int(11)</td><td></td><td>Determines which one of the xrefs will be used as the gene name.</td><td></td></tr>
      <tr class="bg2"><td><b>db_display_name</b></td><td>varchar(255)</td><td>NULL</td><td>Database display name</td><td></td></tr>
      <tr class="bg1"><td><b>type</b></td><td>enum('ARRAY','ALT_TRANS','ALT_GENE','MISC','LIT','PRIMARY_DB_SYNONYM','ENSEMBL')</td><td>NULL</td><td>Type, e.g. 'ARRAY', 'ALT_TRANS', 'ALT_GENE', 'MISC', 'LIT', 'PRIMARY_DB_SYNONYM', 'ENSEMBL'</td><td></td></tr>
      <tr class="bg2"><td><b>secondary_db_name</b></td><td>varchar(255)</td><td>NULL</td><td>Secondary database name</td><td></td></tr>
      <tr class="bg1"><td><b>secondary_db_table</b></td><td>varchar(255)</td><td>NULL</td><td>Secondary database table</td><td></td></tr>
      <tr class="bg2"><td><b>description</b></td><td>text</td><td></td><td>Description of the external database</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#member_xref">member_xref</a></li>
  </ul>


  <div id="member_xref" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">member_xref</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_member_xref" style="cursor:pointer;text-decoration:none" onclick="show_hide('member_xref')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores cross-references for gene members derived from the core databases. It is used by Bio::EnsEMBL::Compara::DBSQL::XrefMemberAdaptor and provides the data used in highlighting gene trees by GO and InterPro annotation" </p>

  <div id="div_member_xref" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>gene_member_id</b></td><td>int(10)</td><td></td><td>External reference to gene_member_id in the <a href="#gene_member">gene_member</a> table. Indicates the gene to which the xref applies.</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>dbprimary_acc</b></td><td>varchar(10)</td><td></td><td>Accession of xref (e.g. GO term, InterPro accession)</td><td>primary key</td></tr>
      <tr class="bg1"><td><b>external_db_id</b></td><td>int(10)</td><td></td><td>External reference to external_db_id in the <a href="#external_db">external_db</a> table. Indicates to which external database the xref belongs.</td><td>primary key</td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#external_db">external_db</a></li>
    <li><a href="#gene_member">gene_member</a></li>
  </ul>


  <div id="other_member_sequence" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">other_member_sequence</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_other_member_sequence" style="cursor:pointer;text-decoration:none" onclick="show_hide('other_member_sequence')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table includes alternative sequences for Member, like sequences with flanking regions</p>

  <div id="div_other_member_sequence" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>seq_type</b></td><td>VARCHAR(40)</td><td></td><td>A short description of this alternative sequence</td><td>primary key</td></tr>
      <tr class="bg1"><td><b>length</b></td><td>int(10)</td><td></td><td>The length of this sequence</td><td></td></tr>
      <tr class="bg2"><td><b>sequence</b></td><td>mediumtext</td><td></td><td>The actual sequence</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#sequence">sequence</a></li>
  </ul>


  <div id="peptide_align_feature" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">peptide_align_feature</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_peptide_align_feature" style="cursor:pointer;text-decoration:none" onclick="show_hide('peptide_align_feature')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">: This table stores the raw local alignment results of peptide to peptide alignments returned by a BLAST run. The hits are actually stored in species-specific tables rather than in a single table. For example, human has the genome_db_id 150, and all the hits that have a human gene as a query are stored in peptide_align_feature_150</p>

  <div id="div_peptide_align_feature" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>peptide_align_feature_id</b></td><td>bigint</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>qmember_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table for the query peptide</td><td></td></tr>
      <tr class="bg1"><td><b>hmember_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table for the hit peptide</td><td></td></tr>
      <tr class="bg2"><td><b>qgenome_db_id</b></td><td>int(10)</td><td></td><td>External reference to genome_db_id in the <a href="#genome_db">genome_db</a> table for the query peptide (for query optimization)</td><td></td></tr>
      <tr class="bg1"><td><b>hgenome_db_id</b></td><td>int(10)</td><td></td><td>External reference to genome_db_id in the <a href="#genome_db">genome_db</a> table for the hit peptide (for query optimization)</td><td></td></tr>
      <tr class="bg2"><td><b>qstart</b></td><td>int(10)</td><td>0</td><td>Starting position in the query peptide sequence</td><td></td></tr>
      <tr class="bg1"><td><b>qend</b></td><td>int(10)</td><td>0</td><td>Ending position in the query peptide sequence</td><td></td></tr>
      <tr class="bg2"><td><b>hstart</b></td><td>int(11)</td><td>0</td><td>Starting position in the hit peptide sequence</td><td></td></tr>
      <tr class="bg1"><td><b>hend</b></td><td>int(11)</td><td>0</td><td>Ending position in the hit peptide sequence</td><td></td></tr>
      <tr class="bg2"><td><b>score</b></td><td>double(16,4)</td><td>0.0000</td><td>Blast score for this HSP</td><td></td></tr>
      <tr class="bg1"><td><b>evalue</b></td><td>double</td><td></td><td>Blast evalue for this HSP</td><td></td></tr>
      <tr class="bg2"><td><b>align_length</b></td><td>int(10)</td><td></td><td>Alignment length of HSP</td><td></td></tr>
      <tr class="bg1"><td><b>identical_matches</b></td><td>int(10)</td><td></td><td>Blast HSP match score</td><td></td></tr>
      <tr class="bg2"><td><b>perc_ident</b></td><td>int(10)</td><td></td><td>Percent identical matches in the HSP length</td><td></td></tr>
      <tr class="bg1"><td><b>positive_matches</b></td><td>int(10)</td><td></td><td>Blast HSP positive score</td><td></td></tr>
      <tr class="bg2"><td><b>perc_pos</b></td><td>int(10)</td><td></td><td>Percent positive matches in the HSP length</td><td></td></tr>
      <tr class="bg1"><td><b>hit_rank</b></td><td>int(10)</td><td></td><td>Rank in blast result</td><td></td></tr>
      <tr class="bg2"><td><b>cigar_line</b></td><td>mediumtext</td><td></td><td>Cigar string coding the actual alignment</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">Example of peptide_align_feature entry: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM peptide_align_feature_150 WHERE hgenome_db_id = 111 LIMIT 1;</pre> <a id="e_peptide_align_feature1" style="cursor:pointer;text-decoration:none" onclick="show_hide('peptide_align_feature1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_peptide_align_feature1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>peptide_align_feature_id</th><th>qmember_id</th><th>hmember_id</th><th>qgenome_db_id</th><th>hgenome_db_id</th><th>qstart</th><th>qend</th><th>hstart</th><th>hend</th><th>score</th><th>evalue</th><th>align_length</th><th>identical_matches</th><th>perc_ident</th><th>positive_matches</th><th>perc_pos</th><th>hit_rank</th><th>cigar_line</th></tr>      <tr><td>15000058332</td><td>9604488</td><td>6982469</td><td>150</td><td>111</td><td>7</td><td>463</td><td>2</td><td>464</td><td>1183.0000</td><td>5e-158</td><td>468</td><td>234</td><td>50</td><td>291</td><td>62</td><td>1</td><td>13M7D9MI79M3I98M2D124MD32MD59MI38M</td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">The following query corresponds to a particular hit found between a Homo sapiens protein and a Anolis carolinensis protein: </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT g1.name as qgenome, m1.stable_id as qstable_id, g2.name as hgenome, m2.stable_id as hstable_id, score, evalue FROM peptide_align_feature_150 JOIN seq_member m1 ON (qmember_id = m1.seq_member_id) JOIN seq_member m2 ON (hmember_id = m2.seq_member_id) JOIN genome_db g1 ON (qgenome_db_id = g1.genome_db_id) JOIN genome_db g2 ON (hgenome_db_id = g2.genome_db_id) WHERE hgenome_db_id = 111 LIMIT 1;</pre> <a id="e_peptide_align_feature2" style="cursor:pointer;text-decoration:none" onclick="show_hide('peptide_align_feature2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_peptide_align_feature2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>qgenome</th><th>qstable_id</th><th>hgenome</th><th>hstable_id</th><th>score</th><th>evalue</th></tr>      <tr><td>homo_sapiens</td><td>ENSP00000247153</td><td>anolis_carolinensis</td><td>ENSACAP00000010758</td><td>1183.0000</td><td>5e-158</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#genome_db">genome_db</a></li>
  </ul>


  <div id="family" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">family</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_family" style="cursor:pointer;text-decoration:none" onclick="show_hide('family')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains all the group homologies found. There are several family_member entries for each family entry.</p>

  <div id="div_family" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>family_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>stable_id</b></td><td>varchar(40)</td><td></td><td>Stable family ID. NOTE: stable_id are currently not stable. We are working in getting IDs stable between releases.</td><td>unique: stable_id</td></tr>
      <tr class="bg1"><td><b>version</b></td><td>INT</td><td></td><td>Version of the stable_id (may only change when members migrate between this family and another one; stays the same otherwise)</td><td></td></tr>
      <tr class="bg2"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_species_set_id in the method_link_species_set table</td><td>key: method_link_species_set_id</td></tr>
      <tr class="bg1"><td><b>description</b></td><td>varchar(255)</td><td></td><td>Description of the family as found using the Longest Common String (LCS) of the descriptions of the member proteins.</td><td>key: description</td></tr>
      <tr class="bg2"><td><b>description_score</b></td><td>double</td><td></td><td>Scores the accuracy of the annotation (max. 100)</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query retrieves families with "CATHELICIDIN" description and description_score of 100 </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM family WHERE description like '%CATHELICIDIN%' AND description_score = 100;</pre> <a id="e_family" style="cursor:pointer;text-decoration:none" onclick="show_hide('family','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_family" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>family_id</th><th>stable_id</th><th>version</th><th>method_link_species_set_id</th><th>description</th><th>description_score</th></tr>      <tr><td>41723</td><td>ENSFM00350000109081</td><td>2</td><td>30047</td><td>CATHELICIDIN {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr class="bg2"><td>45171</td><td>ENSFM00710001442147</td><td>1</td><td>30047</td><td>CATHELICIDIN PRECURSOR CATH FOWLICIDIN</td><td>100</td></tr>      <tr><td>59226</td><td>ENSFM00540000722157</td><td>1</td><td>30047</td><td>CATHELICIDIN ANTIMICROBIAL PEPTIDE PRECURSOR CATHELICIDIN RELATED</td><td>100</td></tr>      <tr class="bg2"><td>87609</td><td>ENSFM00500000314317</td><td>1</td><td>30047</td><td>CATHELICIDIN ANTIMICROBIAL PEPTIDE {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr><td>100034</td><td>ENSFM00500000451410</td><td>4</td><td>30047</td><td>CATHELICIDIN {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr class="bg2"><td>119297</td><td>ENSFM00640001106201</td><td>2</td><td>30047</td><td>CATHELICIDIN 3 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr><td>121739</td><td>ENSFM00310000089953</td><td>1</td><td>30047</td><td>CATHELICIDIN B1 PRECURSOR CATH B1</td><td>100</td></tr>      <tr class="bg2"><td>151560</td><td>ENSFM00500000376284</td><td>1</td><td>30047</td><td>CATHELICIDIN 3 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr><td>151561</td><td>ENSFM00550000766356</td><td>1</td><td>30047</td><td>CATHELICIDIN {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr class="bg2"><td>151562</td><td>ENSFM00730001529592</td><td>1</td><td>30047</td><td>CATHELICIDIN ANTIMICROBIAL PEPTIDE {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr><td>151828</td><td>ENSFM00750001650204</td><td>1</td><td>30047</td><td>CATHELICIDIN {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr class="bg2"><td>209635</td><td>ENSFM00700001402163</td><td>1</td><td>30047</td><td>CATHELICIDIN 2 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr><td>209637</td><td>ENSFM00660001176414</td><td>1</td><td>30047</td><td>CATHELICIDIN ANTIMICROBIAL PEPTIDE {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr class="bg2"><td>214259</td><td>ENSFM00520000549325</td><td>1</td><td>30047</td><td>CATHELICIDIN 1 {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr><td>403139</td><td>ENSFM00760001750198</td><td>1</td><td>30047</td><td>CATHELICIDIN 2 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr class="bg2"><td>405656</td><td>ENSFM00720001501210</td><td>1</td><td>30047</td><td>CATHELICIDIN ANTIMICROBIAL PEPTIDE {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr><td>405657</td><td>ENSFM00800002016908</td><td>1</td><td>30047</td><td>CATHELICIDIN 2 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr class="bg2"><td>407970</td><td>ENSFM00800002016909</td><td>1</td><td>30047</td><td>CATHELICIDIN 2 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr><td>407971</td><td>ENSFM00760001752539</td><td>1</td><td>30047</td><td>CATHELICIDIN {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr class="bg2"><td>436253</td><td>ENSFM00690001368966</td><td>1</td><td>30047</td><td>CATHELICIDIN AL {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr><td>442475</td><td>ENSFM00700001405472</td><td>1</td><td>30047</td><td>CATHELICIDIN 1 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr class="bg2"><td>449219</td><td>ENSFM00780001914357</td><td>1</td><td>30047</td><td>CATHELICIDIN B1 {ECO: |EMBL: 1} FRAGMENT</td><td>100</td></tr>      <tr><td>679584</td><td>ENSFM00660001186067</td><td>1</td><td>30047</td><td>CATHELICIDIN {ECO: |EMBL: 1}</td><td>100</td></tr>      <tr class="bg2"><td>745634</td><td>ENSFM00760001768299</td><td>1</td><td>30047</td><td>CATHELICIDIN  PEPTIDE SHORT 1 {ECO: |EMBL: 1}</td><td>100</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
  </ul>


  <div id="family_member" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">family_member</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_family_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('family_member')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the proteins corresponding to protein family relationship found. There are several family_member entries for each family entry</p>

  <div id="div_family_member" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>family_id</b></td><td>int(10)</td><td></td><td>External reference to family_id in the <a href="#family">family</a> table</td><td>primary key<br />key: family_id</td></tr>
      <tr class="bg2"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>External reference to the seq_member_id in the <a href="#seq_member">seq_member</a> table</td><td>primary key<br />key: seq_member_id</td></tr>
      <tr class="bg1"><td><b>cigar_line</b></td><td>mediumtext</td><td></td><td>Internal description of the multiple alignment (see the description in the <a href="#homology_member">homology_member</a> table)</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query refers to the members of the protein family ENSFM00500000300962. The proteins can be retieved using the member_ids. The multiple alignment can be restored using the cigar_lines. </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT family_member.* FROM family_member JOIN family USING (family_id) WHERE stable_id = "ENSFM00500000300962";</pre> <a id="e_family_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('family_member','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_family_member" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>family_id</th><th>seq_member_id</th><th>cigar_line</th></tr>      <tr><td>49882</td><td>8555714</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>200091832</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>202184619</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>202845003</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>203287871</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>203291435</td><td>49D207M24D</td></tr>      <tr><td>49882</td><td>203968348</td><td>49D8M3D47MD165M7D</td></tr>      <tr class="bg2"><td>49882</td><td>204443459</td><td>49D8M3D47MD32MD133M6D</td></tr>      <tr><td>49882</td><td>204564975</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>204911267</td><td>92M6D3M72D2M38D3M64D</td></tr>      <tr><td>49882</td><td>205655652</td><td>274M6D</td></tr>      <tr class="bg2"><td>49882</td><td>205672632</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>205723761</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>205730410</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>205862909</td><td>49D8M3D202M5D13M</td></tr>      <tr class="bg2"><td>49882</td><td>205977582</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>206190254</td><td>35D239M6D</td></tr>      <tr class="bg2"><td>49882</td><td>206334139</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>206366443</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>206390533</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>206448736</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>206923735</td><td>49D225M6D</td></tr>      <tr><td>49882</td><td>207151056</td><td>49D225M6D</td></tr>      <tr class="bg2"><td>49882</td><td>207258389</td><td>49D225M6D</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#family">family</a></li>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#homology_member">homology_member</a></li>
  </ul>


  <div id="gene_align" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_align</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_align" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_align')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores information about alignments for members</p>

  <div id="div_gene_align" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>gene_align_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>seq_type</b></td><td>varchar(40)</td><td></td><td>Short description of this alignment</td><td></td></tr>
      <tr class="bg1"><td><b>aln_method</b></td><td>varchar(40)</td><td>''</td><td>The alignment method used</td><td></td></tr>
      <tr class="bg2"><td><b>aln_length</b></td><td>int(10)</td><td>0</td><td>The total length of the alignment</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#gene_align_member">gene_align_member</a></li>
  </ul>


  <div id="gene_align_member" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_align_member</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_align_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_align_member')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table allows certain nodes (leaves) to have aligned protein member_scores attached to them</p>

  <div id="div_gene_align_member" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>gene_align_id</b></td><td>int(10)</td><td></td><td>External reference to gene_align_id in the <a href="#gene_align">gene_align</a> table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table in many-to-1 relation (single member per node)</td><td>primary key<br />key: seq_member_id</td></tr>
      <tr class="bg1"><td><b>cigar_line</b></td><td>mediumtext</td><td></td><td>String with the alignment score values</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#gene_align">gene_align</a></li>
  </ul>


  <div id="gene_tree_node" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_tree_node</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_tree_node" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_node')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table holds the gene tree data structure, such as root, relation between parent and child, leaves, etc... In our data structure, all the trees of a given clusterset are arbitrarily connected to the same root. This eases to store and query in the same database the data from independant tree building analysis. Hence the "biological roots" of the trees are the children nodes of the main clusterset root. See the examples below.</p>

  <div id="div_gene_tree_node" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>node_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>parent_id</b></td><td>int(10)</td><td></td><td>Link to the parent node</td><td>key: parent_id</td></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>int(10)</td><td></td><td>Link to the root node</td><td>key: root_id<br />key: root_id_left_index</td></tr>
      <tr class="bg2"><td><b>left_index</b></td><td>int(10)</td><td>0</td><td>Internal index. See above</td><td>key: root_id_left_index</td></tr>
      <tr class="bg1"><td><b>right_index</b></td><td>int(10)</td><td>0</td><td>Internal index. See above</td><td></td></tr>
      <tr class="bg2"><td><b>distance_to_parent</b></td><td>double</td><td>1.0</td><td>Phylogenetic distance between this node and its parent</td><td></td></tr>
      <tr class="bg1"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table to allow linkage from trees to peptides/transcripts.</td><td>key: seq_member_id</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query returns the root nodes of the independant protein trees stored in the database </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT gtn.node_id FROM gene_tree_node gtn LEFT JOIN gene_tree_root gtr ON (gtn.parent_id = gtr.root_id) WHERE gtr.tree_type = 'clusterset' AND gtr.member_type = 'protein' LIMIT 10;</pre> <a id="e_gene_tree_node" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_node','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_gene_tree_node" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>gtn.node_id</th></tr>      <tr><td>2</td></tr>      <tr class="bg2"><td>1298</td></tr>      <tr><td>2578</td></tr>      <tr class="bg2"><td>3848</td></tr>      <tr><td>5078</td></tr>      <tr class="bg2"><td>6273</td></tr>      <tr><td>7450</td></tr>      <tr class="bg2"><td>8618</td></tr>      <tr><td>9786</td></tr>      <tr class="bg2"><td>10936</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#gene_tree_root">gene_tree_root</a></li>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#gene_tree_node_attr">gene_tree_node_attr</a></li>
    <li><a href="#gene_tree_node_tag">gene_tree_node_tag</a></li>
  </ul>


  <div id="gene_tree_root" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_tree_root</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_tree_root" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_root')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">Header table for gene_trees. The database is able to contain several sets of trees computed on the same genes. We call these analysis "clustersets" and they can be distinguished with the clusterset_id field. Traditionally, the compara databases have contained only one clusterset (clusterset_id=1), but currently (starting on release 66) we have at least 2 (one for protein trees and one for ncRNA trees). See the examples below.</p>

  <div id="div_gene_tree_root" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>INT(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>member_type</b></td><td>ENUM('protein', 'ncrna')</td><td></td><td>The type of members used in the tree</td><td></td></tr>
      <tr class="bg1"><td><b>tree_type</b></td><td>ENUM('clusterset', 'supertree', 'tree')</td><td></td><td>The type of the tree</td><td>key: tree_type</td></tr>
      <tr class="bg2"><td><b>clusterset_id</b></td><td>VARCHAR(20)</td><td>'default'</td><td>Name for the set of clusters/trees</td><td></td></tr>
      <tr class="bg1"><td><b>method_link_species_set_id</b></td><td>INT(10)</td><td></td><td>External reference to method_link_species_set_id in the method_link_species_set table</td><td></td></tr>
      <tr class="bg2"><td><b>gene_align_id</b></td><td>INT(10)</td><td></td><td>External reference to gene_align_id in the <a href="#gene_align">gene_align</a> table</td><td></td></tr>
      <tr class="bg1"><td><b>ref_root_id</b></td><td>INT(10)</td><td></td><td>External reference to default (merged) root_id for this tree</td><td>key: ref_root_id</td></tr>
      <tr class="bg2"><td><b>stable_id</b></td><td>VARCHAR(40)</td><td></td><td>Unique, stable ID for the tree (follows the pattern: label(5).release_introduced(4).unique_id(10))</td><td>unique: key</td></tr>
      <tr class="bg1"><td><b>version</b></td><td>INT</td><td></td><td>Version of the stable ID (changes only when members move to/from existing trees)</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">The following query retrieves all the node_id of the current clustersets </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT * FROM gene_tree_root WHERE tree_type = 'clusterset';</pre> <a id="e_gene_tree_root1" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_root1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_gene_tree_root1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>root_id</th><th>member_type</th><th>tree_type</th><th>clusterset_id</th><th>method_link_species_set_id</th><th>gene_align_id</th><th>ref_root_id</th><th>stable_id</th><th>version</th></tr>      <tr><td>100000001</td><td>ncrna</td><td>clusterset</td><td>default</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204624</td><td>ncrna</td><td>clusterset</td><td>pg_it_nj</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204625</td><td>ncrna</td><td>clusterset</td><td>ml_it_10</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204626</td><td>ncrna</td><td>clusterset</td><td>pg_it_phyml</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204627</td><td>ncrna</td><td>clusterset</td><td>ss_it_s16</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204628</td><td>ncrna</td><td>clusterset</td><td>ss_it_s6a</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204629</td><td>ncrna</td><td>clusterset</td><td>ss_it_s16a</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204630</td><td>ncrna</td><td>clusterset</td><td>ss_it_s6b</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204631</td><td>ncrna</td><td>clusterset</td><td>ss_it_s16b</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204632</td><td>ncrna</td><td>clusterset</td><td>ss_it_s6c</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204633</td><td>ncrna</td><td>clusterset</td><td>ss_it_s6d</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204634</td><td>ncrna</td><td>clusterset</td><td>ss_it_s6e</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204635</td><td>ncrna</td><td>clusterset</td><td>ss_it_s7a</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204636</td><td>ncrna</td><td>clusterset</td><td>ss_it_s7b</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204637</td><td>ncrna</td><td>clusterset</td><td>ss_it_s7c</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204638</td><td>ncrna</td><td>clusterset</td><td>ss_it_s7d</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204639</td><td>ncrna</td><td>clusterset</td><td>ss_it_s7e</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204640</td><td>ncrna</td><td>clusterset</td><td>ss_it_s7f</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204641</td><td>ncrna</td><td>clusterset</td><td>ft_it_ml</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204642</td><td>ncrna</td><td>clusterset</td><td>ft_it_nj</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>100204643</td><td>ncrna</td><td>clusterset</td><td>ftga_it_ml</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>100204644</td><td>ncrna</td><td>clusterset</td><td>ftga_it_nj</td><td>40102</td><td></td><td></td><td></td><td></td></tr>      <tr><td>1</td><td>protein</td><td>clusterset</td><td>default</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>1262340</td><td>protein</td><td>clusterset</td><td>treebest</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr><td>1262341</td><td>protein</td><td>clusterset</td><td>phyml-aa</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>1262342</td><td>protein</td><td>clusterset</td><td>phyml-nt</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr><td>1262343</td><td>protein</td><td>clusterset</td><td>nj-dn</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>1262344</td><td>protein</td><td>clusterset</td><td>nj-ds</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr><td>1262345</td><td>protein</td><td>clusterset</td><td>nj-mm</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>1262346</td><td>protein</td><td>clusterset</td><td>raxml</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr><td>1262347</td><td>protein</td><td>clusterset</td><td>raxml_bl</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>1262348</td><td>protein</td><td>clusterset</td><td>notung</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr><td>1262349</td><td>protein</td><td>clusterset</td><td>copy</td><td>40101</td><td></td><td></td><td></td><td></td></tr>      <tr class="bg2"><td>1262350</td><td>protein</td><td>clusterset</td><td>raxml_update</td><td>40101</td><td></td><td></td><td></td><td></td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">To get the number of trees of each type </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT member_type, tree_type, COUNT(*) FROM gene_tree_root GROUP BY member_type, tree_type;</pre> <a id="e_gene_tree_root2" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_root2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_gene_tree_root2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>member_type</th><th>tree_type</th><th>COUNT(*)</th></tr>      <tr><td>protein</td><td>clusterset</td><td>12</td></tr>      <tr class="bg2"><td>protein</td><td>tree</td><td>92693</td></tr>      <tr><td>ncrna</td><td>clusterset</td><td>22</td></tr>      <tr class="bg2"><td>ncrna</td><td>supertree</td><td>25</td></tr>      <tr><td>ncrna</td><td>tree</td><td>18180</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#gene_tree_node">gene_tree_node</a></li>
    <li><a href="#gene_tree_root_tag">gene_tree_root_tag</a></li>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
    <li><a href="#gene_align">gene_align</a></li>
  </ul>


  <div id="gene_tree_node_tag" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_tree_node_tag</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_tree_node_tag" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_node_tag')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains several tag/value data attached to node_ids</p>

  <div id="div_gene_tree_node_tag" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>node_id</b></td><td>int(10)</td><td></td><td>External reference to node_id in the <a href="#gene_tree_node">gene_tree_node</a> table</td><td>key: node_id_tag<br />key: node_id</td></tr>
      <tr class="bg2"><td><b>tag</b></td><td>varchar(50)</td><td></td><td>Tag name for the tag/value pair</td><td>key: node_id_tag</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>mediumtext</td><td></td><td>Value for the tag/value pair</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#gene_tree_node">gene_tree_node</a></li>
    <li><a href="#gene_tree_node_attr">gene_tree_node_attr</a></li>
  </ul>


  <div id="gene_tree_root_tag" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_tree_root_tag</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_tree_root_tag" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_root_tag')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains several tag/value data for gene_tree_roots</p>

  <div id="div_gene_tree_root_tag" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>root_id</b></td><td>int(10)</td><td></td><td>External reference to root_id in the <a href="#gene_tree_root">gene_tree_root</a> table</td><td>key: root_id_tag<br />key: root_id</td></tr>
      <tr class="bg2"><td><b>tag</b></td><td>varchar(50)</td><td></td><td>Tag name for the tag/value pair</td><td>key: root_id_tag</td></tr>
      <tr class="bg1"><td><b>value</b></td><td>mediumtext</td><td></td><td>Value for the tag/value pair</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#gene_tree_root">gene_tree_root</a></li>
  </ul>


  <div id="gene_tree_node_attr" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">gene_tree_node_attr</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_gene_tree_node_attr" style="cursor:pointer;text-decoration:none" onclick="show_hide('gene_tree_node_attr')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains several gene tree attributes data attached to node_ids</p>

  <div id="div_gene_tree_node_attr" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>node_id</b></td><td>INT(10)</td><td></td><td>External reference to node_id in the <a href="#gene_tree_node">gene_tree_node</a> table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>node_type</b></td><td>ENUM("duplication", "dubious", "speciation", "gene_split")</td><td></td><td>Type of homology</td><td></td></tr>
      <tr class="bg1"><td><b>species_tree_node_id</b></td><td>INT(10)</td><td></td><td>Taxon / Ancestral species annotation (given as the node_id in the species tree)</td><td></td></tr>
      <tr class="bg2"><td><b>bootstrap</b></td><td>TINYINT</td><td></td><td>The bootstrap value of the node in the tree</td><td></td></tr>
      <tr class="bg1"><td><b>duplication_confidence_score</b></td><td>DOUBLE(5,4)</td><td></td><td>The calculated confidence score for duplications</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#gene_tree_node">gene_tree_node</a></li>
    <li><a href="#gene_tree_node_tag">gene_tree_node_tag</a></li>
  </ul>


  <div id="hmm_profile" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">hmm_profile</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_hmm_profile" style="cursor:pointer;text-decoration:none" onclick="show_hide('hmm_profile')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores different HMM-based profiles used and produced by gene trees</p>

  <div id="div_hmm_profile" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>model_id</b></td><td>varchar(40)</td><td></td><td>Model ID of the profile. This is the stable_id of the gene-tree, or the external ID in case of imported models (such as RF00001)</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>name</b></td><td>varchar(40)</td><td></td><td>Name of the model, if available (such as 5S_rRNA for RF00001)</td><td></td></tr>
      <tr class="bg1"><td><b>type</b></td><td>varchar(40)</td><td></td><td>Short description of the origin of the profile (usually, one of "tree_hmm_aa_v3", "tree_hmm_dna_v3", "infernal" or "infernal-refined")</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>compressed_profile</b></td><td>mediumblob</td><td></td><td>The HMM profile, compressed with zlib. It can be decompressed with the MySQL function UNCOMPRESS()</td><td></td></tr>
      <tr class="bg1"><td><b>consensus</b></td><td>mediumtext</td><td></td><td>The consensus sequence derived from the profile</td><td></td></tr>
    </table>
  </div>

  <div id="hmm_annot" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">hmm_annot</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_hmm_annot" style="cursor:pointer;text-decoration:none" onclick="show_hide('hmm_annot')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores the HMM annotation of the seq_members</p>

  <div id="div_hmm_annot" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>External reference to a seq_member_id in the <a href="#seq_member">seq_member</a> table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>model_id</b></td><td>varchar(40)</td><td>NULL</td><td>External reference to the internal numeric ID of a HMM profile in <a href="#hmm_profile">hmm_profile</a></td><td>key: model_id</td></tr>
      <tr class="bg1"><td><b>evalue</b></td><td>float</td><td></td><td>The e-value of the hit</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#hmm_profile">hmm_profile</a></li>
  </ul>


  <div id="hmm_curated_annot" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">hmm_curated_annot</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_hmm_curated_annot" style="cursor:pointer;text-decoration:none" onclick="show_hide('hmm_curated_annot')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores the curated / forced HMM annotation of the seq_members</p>

  <div id="div_hmm_curated_annot" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>seq_member_stable_id</b></td><td>varchar(40)</td><td></td><td>External reference to a seq_member_id in the <a href="#seq_member">seq_member</a> table</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>model_id</b></td><td>varchar(40)</td><td>NULL</td><td>External reference to the internal numeric ID of a HMM profile in <a href="#hmm_profile">hmm_profile</a></td><td>key: model_id</td></tr>
      <tr class="bg1"><td><b>library_version</b></td><td>varchar(40)</td><td></td><td>Name of the HMM library against the curation has been done</td><td></td></tr>
      <tr class="bg2"><td><b>annot_date</b></td><td>TIMESTAMP</td><td>CURRENT_TIMESTAMP</td><td>When did the curation happened</td><td></td></tr>
      <tr class="bg1"><td><b>reason</b></td><td>MEDIUMTEXT</td><td></td><td>Why are we forcing this curation</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#hmm_profile">hmm_profile</a></li>
  </ul>


  <div id="homology" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">homology</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_homology" style="cursor:pointer;text-decoration:none" onclick="show_hide('homology')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains all the genomic homologies. There are two homology_member entries for each homology entry for now, but both the schema and the API can handle more than just pairwise relationships. <br />dN, dS, N, S and lnL are statistical values given by the codeml program of the <a href="http://abacus.gene.ucl.ac.uk/software/paml.html">Phylogenetic Analysis by Maximum Likelihood (PAML)</a> package.</p>

  <div id="div_homology" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>homology_id</b></td><td>int(10)</td><td></td><td>Unique internal ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>method_link_species_set_id</b></td><td>int(10)</td><td></td><td>External reference to method_link_species_set_id in the <a href="#method_link_species_set">method_link_species_set</a> table</td><td>key: method_link_species_set_id</td></tr>
      <tr class="bg1"><td><b>description</b></td><td>ENUM('ortholog_one2one','ortholog_one2many','ortholog_many2many','within_species_paralog','other_paralog','gene_split','between_species_paralog','alt_allele','homoeolog_one2one','homoeolog_one2many','homoeolog_many2many')</td><td></td><td>A normalized, short description of the homology relationship</td><td></td></tr>
      <tr class="bg2"><td><b>is_tree_compliant</b></td><td>tinyint(1)</td><td>0</td><td>Whether the homology is fully compliant with the tree and the definition of orthology / paralogy</td><td></td></tr>
      <tr class="bg1"><td><b>dn</b></td><td>float(10,5)</td><td></td><td>The non-synonymous mutation rate</td><td></td></tr>
      <tr class="bg2"><td><b>ds</b></td><td>float(10,5)</td><td></td><td>The synonymous mutation rate</td><td></td></tr>
      <tr class="bg1"><td><b>n</b></td><td>float(10,1)</td><td></td><td>The estimated number of non-synonymous mutations</td><td></td></tr>
      <tr class="bg2"><td><b>s</b></td><td>float(10,1)</td><td></td><td>The estimated number of synonymous mutations</td><td></td></tr>
      <tr class="bg1"><td><b>lnl</b></td><td>float(10,3)</td><td></td><td>The negative log likelihood of the estimation</td><td></td></tr>
      <tr class="bg2"><td><b>species_tree_node_id</b></td><td>int(10)</td><td></td><td>The node_id of the species-tree node to which the homology is attached</td><td>key: species_tree_node_id</td></tr>
      <tr class="bg1"><td><b>gene_tree_node_id</b></td><td>int(10)</td><td></td><td>The node_id of the gene-tree node from which the homology is derived</td><td>key: gene_tree_node_id</td></tr>
      <tr class="bg2"><td><b>gene_tree_root_id</b></td><td>int(10)</td><td></td><td>The root_id of the gene tree from which the homology is derived</td><td>key: gene_tree_root_id</td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example 1:</p><p style="padding-left:10px">The following query defines a pair of paralogous xenopous genes. See <a href="#homology_member">homology_member</a> for more details </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT homology.* FROM homology JOIN method_link_species_set USING (method_link_species_set_id) WHERE name="X.tro paralogues" LIMIT 1;</pre> <a id="e_homology1" style="cursor:pointer;text-decoration:none" onclick="show_hide('homology1','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_homology1" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>homology_id</th><th>method_link_species_set_id</th><th>description</th><th>is_tree_compliant</th><th>dn</th><th>ds</th><th>n</th><th>s</th><th>lnl</th><th>species_tree_node_id</th><th>gene_tree_node_id</th><th>gene_tree_root_id</th></tr>      <tr><td>100012483</td><td>27763</td><td>other_paralog</td><td>1</td><td></td><td></td><td></td><td></td><td></td><td>40102006</td><td>100150796</td><td>100150796</td></tr>    </table>
  </div></div><div style="margin: 10px 5px"><p style="font-weight:bold">Example 2:</p><p style="padding-left:10px">See species_names that participate in this particular homology entry </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT homology_id, description, GROUP_CONCAT(genome_db.name) AS species FROM homology JOIN method_link_species_set USING (method_link_species_set_id) JOIN species_set USING (species_set_id) JOIN genome_db USING(genome_db_id) WHERE method_link_id=201 AND homology_id<5000000  GROUP BY homology_id LIMIT 4;</pre> <a id="e_homology2" style="cursor:pointer;text-decoration:none" onclick="show_hide('homology2','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_homology2" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>homology_id</th><th>description</th><th>species</th></tr>      <tr><td>1</td><td>ortholog_one2one</td><td>astyanax_mexicanus,danio_rerio</td></tr>      <tr class="bg2"><td>2</td><td>ortholog_one2one</td><td>xiphophorus_maculatus,poecilia_formosa</td></tr>      <tr><td>3</td><td>ortholog_one2one</td><td>oryzias_latipes,poecilia_formosa</td></tr>      <tr class="bg2"><td>4</td><td>ortholog_one2one</td><td>oryzias_latipes,xiphophorus_maculatus</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#homology_member">homology_member</a></li>
    <li><a href="#method_link_species_set">method_link_species_set</a></li>
  </ul>


  <div id="homology_member" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">homology_member</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_homology_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('homology_member')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains the sequences corresponding to every genomic homology relationship found. There are two homology_member entries for each pairwise homology entry. As written in the homology table section, both schema and API can deal with more than pairwise relationships.<br /> The original alignment is not stored but it can be retrieved using the cigar_line field and the original sequences. The cigar line defines the sequence of matches or mismatches and deletions in the alignment. <ul> <li><b>First peptide sequence</b>: SERCQVVVISIGPISVLSMILDFY</li> <li><b>Second peptide sequence</b>: SDRCQVLVISILSMIGLDFY</li> <li><b>First corresponding cigar line</b>: 20MD4M</li> <li><b>Second corresponding cigar line</b>: 11M5D9M</li> </ul> The alignment will be:<br /> <table class="spreadsheet"> <caption>Example of alignment reconstruction</caption> <tbody> <tr> <th>First peptide cigar line</th> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>D</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> </tr> <tr> <th>First aligned peptide</th> <td>S</td> <td>E</td> <td>R</td> <td>C</td> <td>Q</td> <td>V</td> <td>V</td> <td>V</td> <td>I</td> <td>S</td> <td>I</td> <td>G</td> <td>P</td> <td>I</td> <td>S</td> <td>V</td> <td>L</td> <td>S</td> <td>M</td> <td>I</td> <td>-</td> <td>L</td> <td>D</td> <td>F</td> <td>Y</td> </tr> <tr> <th>Second aligned peptide</th> <td>S</td> <td>D</td> <td>R</td> <td>C</td> <td>Q</td> <td>V</td> <td>L</td> <td>V</td> <td>I</td> <td>S</td> <td>I</td> <td>-</td> <td>-</td> <td>-</td> <td>-</td> <td>-</td> <td>L</td> <td>S</td> <td>M</td> <td>I</td> <td>G</td> <td>L</td> <td>D</td> <td>F</td> <td>Y</td> </tr> <tr> <th>Second peptide cigar line</th> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>D</td> <td>D</td> <td>D</td> <td>D</td> <td>D</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> <td>M</td> </tr> </tbody> </table></p>

  <div id="div_homology_member" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>homology_id</b></td><td>int(10)</td><td></td><td>External reference to homology_id in the <a href="#homology">homology</a> table</td><td>primary key<br />key: homology_id</td></tr>
      <tr class="bg2"><td><b>gene_member_id</b></td><td>int(10)</td><td></td><td>External reference to gene_member_id in the <a href="#gene_member">gene_member</a> table. Refers to the corresponding "ENSMBLGENE" entry</td><td>primary key<br />key: gene_member_id</td></tr>
      <tr class="bg1"><td><b>seq_member_id</b></td><td>int(10)</td><td></td><td>External reference to seq_member_id in the <a href="#seq_member">seq_member</a> table. Refers to the corresponding "ENSEMBLPEP" entry</td><td>key: seq_member_id</td></tr>
      <tr class="bg2"><td><b>cigar_line</b></td><td>mediumtext</td><td></td><td>An internal description of the alignment. It contains mathces/mismatches (M) and delations (D) and refers to the corresponding seq_member_id sequence</td><td></td></tr>
      <tr class="bg1"><td><b>perc_cov</b></td><td>tinyint</td><td>0</td><td>Defines the percentage of the peptide which has been aligned</td><td></td></tr>
      <tr class="bg2"><td><b>perc_id</b></td><td>tinyint</td><td>0</td><td>Defines the percentage of identity between both homologues</td><td></td></tr>
      <tr class="bg1"><td><b>perc_pos</b></td><td>tinyint</td><td>0</td><td>Defines the percentage of positivity (similarity) between both homologues</td><td></td></tr>
    </table>
  </div>
<div style="margin: 10px 5px"><p style="font-weight:bold">Example:</p><p style="padding-left:10px">The following query refers to the two homologue sequences from the first xenopus' paralogy object. Gene and peptide sequence of the second homologue can retrieved in the same way. </p><pre style="display:inline;border:1px solid #555;padding:2px;margin-right:15px;margin-left:10px">SELECT homology_member.* FROM homology_member JOIN homology USING (homology_id) JOIN method_link_species_set USING (method_link_species_set_id) WHERE name="X.tro paralogues" LIMIT 2;</pre> <a id="e_homology_member" style="cursor:pointer;text-decoration:none" onclick="show_hide('homology_member','example')"><img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/><span style="vertical-align:middle"> Show query results</span></a>

  <div id="ex_homology_member" style="display:none;">
    <table class="ss" style="width:50%;margin-top:20px;margin-left:10px;border-spacing:2px">
      <tr><th>homology_id</th><th>gene_member_id</th><th>seq_member_id</th><th>cigar_line</th><th>perc_cov</th><th>perc_id</th><th>perc_pos</th></tr>      <tr><td>100012483</td><td>100070511</td><td>100070511</td><td>38M4D7M6D44MD</td><td>89</td><td>57</td><td>57</td></tr>      <tr class="bg2"><td>100012483</td><td>100073230</td><td>100073230</td><td>2M3D37M4D50M3DM</td><td>88</td><td>57</td><td>57</td></tr>    </table>
  </div></div>  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#gene_member">gene_member</a></li>
    <li><a href="#seq_member">seq_member</a></li>
    <li><a href="#homology">homology</a></li>
  </ul>


  <div id="mapping_session" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">mapping_session</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_mapping_session" style="cursor:pointer;text-decoration:none" onclick="show_hide('mapping_session')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table contains one entry per stable_id mapping session (either for Families or for Protein Trees), which contains the type, the date of the mapping, and which releases were linked together. A single mapping_session is the event when mapping between two given releases for a particular class type ('family' or 'tree') is loaded. The whole event is thought to happen momentarily at 'when_mapped' (used for sorting in historical order).</p>

  <div id="div_mapping_session" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>mapping_session_id</b></td><td>INT</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>type</b></td><td>ENUM('family', 'tree')</td><td></td><td>Type of stable_ids that were mapped during this session</td><td>unique: key</td></tr>
      <tr class="bg1"><td><b>when_mapped</b></td><td>TIMESTAMP</td><td>CURRENT_TIMESTAMP</td><td>Normally, we use the date of creation of the mapping file being loaded. This prevents the date from chaging even if we accidentally remove the entry and have to re-load it.</td><td></td></tr>
      <tr class="bg2"><td><b>rel_from</b></td><td>INT</td><td></td><td>rel.number from which the stable_ids were mapped during this session. rel_from < rel_to</td><td>unique: key</td></tr>
      <tr class="bg1"><td><b>rel_to</b></td><td>INT</td><td></td><td>rel.number to which the stable_ids were mapped during this session. rel_from < rel_to</td><td>unique: key</td></tr>
      <tr class="bg2"><td><b>prefix</b></td><td>CHAR(4)</td><td></td><td>Prefix</td><td>unique: key</td></tr>
    </table>
  </div>

  <div id="stable_id_history" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">stable_id_history</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_stable_id_history" style="cursor:pointer;text-decoration:none" onclick="show_hide('stable_id_history')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table keeps the history of stable_id changes from one release to another. The primary key 'object' describes a set of members migrating from stable_id_from to stable_id_to. Their volume (related to the 'shared_size' of the new class) is reflected by the fractional 'contribution' field. Since both stable_ids are listed in the primary key, they are not allowed to be NULLs. We shall treat empty strings as NULLs. If stable_id_from is empty, it means these members are newcomers into the new release. If stable_id_to is empty, it means these previously known members are disappearing in the new release. If both neither stable_id_from nor stable_id_to is empty, these members are truly migrating.</p>

  <div id="div_stable_id_history" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>mapping_session_id</b></td><td>INT</td><td></td><td>Reference to mapping_session.mapping_session_id. All the stable_ids of a given mapping should have the same session_id</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>stable_id_from</b></td><td>VARCHAR(40)</td><td>''</td><td>The previous stable ID</td><td>primary key</td></tr>
      <tr class="bg1"><td><b>version_from</b></td><td>INT</td><td>NULL</td><td>The version number of the previous stable ID (specific to each stable ID; not to be confused with the release number)</td><td></td></tr>
      <tr class="bg2"><td><b>stable_id_to</b></td><td>VARCHAR(40)</td><td>''</td><td>The new stable ID</td><td>primary key</td></tr>
      <tr class="bg1"><td><b>version_to</b></td><td>INT</td><td>NULL</td><td>The new version number</td><td></td></tr>
      <tr class="bg2"><td><b>contribution</b></td><td>FLOAT</td><td></td><td>Percentage of of the new object (tree / family) that comes from the previous one</td><td></td></tr>
    </table>
  </div>

  <div id="CAFE_gene_family" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">CAFE_gene_family</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_CAFE_gene_family" style="cursor:pointer;text-decoration:none" onclick="show_hide('CAFE_gene_family')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table holds information about each CAFE gene family</p>

  <div id="div_CAFE_gene_family" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>cafe_gene_family_id</b></td><td>int(10)</td><td></td><td>Internal unique ID</td><td>primary key</td></tr>
      <tr class="bg2"><td><b>root_id</b></td><td>int(10)</td><td></td><td>External reference to root_id in the <a href="#species_tree_root">species_tree_root</a> table</td><td>key: root_id</td></tr>
      <tr class="bg1"><td><b>lca_id</b></td><td>int(10)</td><td></td><td>External reference to the lowest common ancestor node in the <a href="#species_tree_node">species_tree_node</a> table</td><td></td></tr>
      <tr class="bg2"><td><b>gene_tree_root_id</b></td><td>int(10)</td><td></td><td>External reference to root_id in the <a href="#gene_tree_root">gene_tree_root</a> table</td><td>key: gene_tree_root_id</td></tr>
      <tr class="bg1"><td><b>pvalue_avg</b></td><td>double(5,4)</td><td>NULL</td><td>The average pvalue for the gene family as reported by CAFE</td><td></td></tr>
      <tr class="bg2"><td><b>lambdas</b></td><td>varchar(100)</td><td>NULL</td><td>The lambda/s values reported/used by CAFE</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#species_tree_node">species_tree_node</a></li>
    <li><a href="#species_tree_root">species_tree_root</a></li>
    <li><a href="#CAFE_species_gene">CAFE_species_gene</a></li>
    <li><a href="#gene_tree_root">gene_tree_root</a></li>
  </ul>


  <div id="CAFE_species_gene" style="width:820px;height:20px;border: 2px groove #CCC;background-color:#FAFAFF;padding:2px;margin-top:35px;margin-bottom:2px">
        <div style="float:left;padding:0px;height:20px;width:10px;background-color:#1E90FF;margin-right:10px"></div>
    <div style="float:left;text-align:left;font-size:11pt;font-weight:bold">CAFE_species_gene</div>
    <div style="float:right;text-align:right;padding-right:1px">
      <a id="a_CAFE_species_gene" style="cursor:pointer;text-decoration:none" onclick="show_hide('CAFE_species_gene')">
        <img src="/i/16/plus-button.png" style="width:12px;height:12px;vertical-align:middle" alt="show"/>
        <span style="vertical-align:middle">Show columns</span>
      </a> 
      <b> | </b> <a href="#top">[Back to top]</a>
    </div>
  </div>
  <p style="padding:5px 0px;margin-bottom:0px;width:800px">This table stores per <a href="#species_tree_node">species_tree_node</a> information about expansions/contractions of each <a href="#CAFE_gene_family">CAFE_gene_family</a></p>

  <div id="div_CAFE_species_gene" style="display:none">
    <table style="border:1px solid #667aa6;padding:0px;min-width:1000px;max-width:1200px">
      <tr class="center" style="color:#FFFFFF;background-color:#667aa6"><th style="color:#FFF;padding:2px">Column</th><th style="color:#FFF;padding:2px">Type</th><th style="color:#FFF;padding:2px;min-width:80px">Default value</th><th style="color:#FFF;padding:2px;min-width:500px">Description</th><th style="color:#FFF;padding:2px;min-width:100px">Index</th></tr>
      <tr class="bg1"><td><b>cafe_gene_family_id</b></td><td>int(10)</td><td></td><td>External reference to cafe_gene_family_id in the <a href="#CAFE_gene_family">CAFE_gene_family</a> table.</td><td>key: cafe_gene_family_id</td></tr>
      <tr class="bg2"><td><b>node_id</b></td><td>int(10)</td><td></td><td>External reference to node_id in the <a href="#species_tree_node">species_tree_node</a> table</td><td></td></tr>
      <tr class="bg1"><td><b>n_members</b></td><td>int(4)</td><td></td><td>The number of members for the node as reported by CAFE</td><td></td></tr>
      <tr class="bg2"><td><b>pvalue</b></td><td>double(5,4)</td><td>NULL</td><td>The pvalue of the node as reported by CAFE</td><td></td></tr>
    </table>
  </div>
  <p style="margin-top:10px"><b>See also:</b></p>
  <ul>
    <li><a href="#species_tree_node">species_tree_node</a></li>
    <li><a href="#CAFE_gene_family">CAFE_gene_family</a></li>
  </ul>



</body>
</html>
