<html><head><title>Eh?: geneview</title>
    


<li id="OrthologuePrediction">
<b>Orthologue Prediction</b> - a link to the closest putative orthologue genes in other species. 
Links are given between human, mouse and rat with all other species including chimp, chicken, zebrafish, 
<i>Takifugu rubripes</i>, 
<i>Anopheles gambiae</i>,
<i>Drosophila melanogaster</i> and 
<i>Caenorhabditis elegans</i>).
A link at the bottom of this section leads to a corresponding Ensembl 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=multicontigview">MultiContigView</a>' page, which allows display of both genome sequence and annotation information for several species simultaneously. 
By default, all species with orthologues defined in this section are selected. 
Further species could be added or others dropped directly from 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=multicontigview">MultiContigView</a>' pages. 
</p>

<p>
The homologous genes represent the best reciprocal BLAST hits for the
two species (Best Reciprocal Hit: BRH), with additional pairs obtained
by a combination of BLAST and location information (Reciprocal Hit
based on Synteny: RHS) for more closely related species. These
homologues may therefore potentially represent orthologues. </p>

<p>
The chimp/human homologues are obtained from whole genome BLASTz
alignments between human and chimp rather than by reciprocal BLASTP
analysis and are designated DWGA (Derived from Whole Genome
Alignments). This was also the method used to predict the chimp genes,
by mapping the human genes onto the chimp chromosomes. See Description
on the relevant Ensembl Gene Report. </p>

<p>
Clicking on the putative orthologue gene will take you to a 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=geneview">GeneView</a>' display of that gene within the web site for the other species. 
See also the 
'<a href="http://www.ensembl.org/Homo_sapiens/helpview?se=1&amp;kw=syntenyview">SyntenyView</a>' help page. 
Where no homology matches are shown, you may wish to explore the Ensembl protein family (see below). 
</p>

<p>
dN (number of non-synonymous substitutions / number of non-synonymous sites) and 
dS (number of synonymous substitutions / number of synonymous sites) values were generated using the <b>codeml</b> program included in the 
<a href="http://abacus.gene.ucl.ac.uk/software/paml.html" target="external">PAML</a> package 
(<a href="#nb1" name="nh1">1</a>). 
With the parameters we have used, <b>codeml</b> performs pair wise Maximum Likelihood calculations of dN and dS for each set of orthologues. 
We have used the F3x4 codon evolution model 
(<a href="#nb2" name="nh2">2</a>).
This takes into account both the bias deriving from the different
probabilities of transition (T&lt;-&gt;C and A&lt;-&gt;G) versus
transversion (T/C&lt;-&gt;A/G) mutations, and the bias due to different
nucleotide frequencies at the three codon positions. </p>

<p>
dN and dS values are only provided for orthologues from some species
pairs, i.e. human/mouse, human/rat, and mouse/rat. Orthologues for
other species pairs are too divergent for dS to be an accurate measure.
Most synonymous sites will have be subjected to more than one mutation
and ancestral changes cannot be reliably inferred from extant
sequences, (i.e. dS is saturated). </p><p>

</p><p>
Orthologue predictions for human/mouse, human/rat and mouse/rat may not
be perfect. Incorrect assignments will manifest anomalously high dS
values. We have, therefore, applied a cut-off of twice the median value
of all dS for each species pair as the criterion for displaying the
dN/dS ratio. Predicted orthology relationships with dS above this
threshold are likely to be errors. </p>

<p>
The following dS threshold values have been used: 
</p>

<pre>                   dS threshold

   human/mouse        1.26978
   human/rat          1.28408
   mouse/rat          0.40745
</pre>

<p>
Therefore, dN/dS ratios for human/mouse orthologues are only displayed when dS &lt;= 1.26978, otherwise "--". 
</p>

<p>
NB: The dN, dS values for elegans/briggsae were calculated but dS
values were considered as saturated (median = 2.26584, &gt;&gt; 1).
After applying the dS threshold (4.53168), the median and average on
the remaining set were 1.93962 and 2.062 respectively, and comparable
with the data that was published in the <i>Caenorhabditis briggsae</i> genome paper 
(<a href="#nb3" name="nh3">3</a>), average 1.78 (no median was provided in the paper). 
</p>


<h5 id="References">References</h5>

<p>
<a href="#nh1" name="nb1"><b>1</b></a>. 
Yang, Z. (1997)<br>
"PAML: a program package for phylogenetic analysis by maximum likehood."<br>
Comput. Appl. Biosci. 1997 13: 555-556.<br>
</p>

<p>
<a href="#nh2" name="nb2"><b>2</b></a>. 
Goldman, N. &amp; Zang, Y. (1994)<br>
"A codon-based model of nucleotide substitution for protein-coding DNA sequences."<br>
Mol. Biol. Evol. 11, 725-736.<br>
</p>

<p>
<a href="#nh3" name="nb3"><b>3</b></a>. 
Stein, LD et al. (2003)<br>
"The Genome Sequence of <i>Caenorhabditis briggsae</i>: A Platform for Comparative Genomics."
PLoS Biol. 1, 166-192.
</p>
</li>

<li id="ParaloguePrediction">
<p>
<b>Paralogue Prediction</b> - a link to putative paralogues. 
<br>
The identification of paralogues (recently duplicated human genes) has
been conducted using a simple approach and it is only available for
human at the moment. <br>
A raw set of gene homologies was derived from an all-<i>versus</i>-all
BLAST approach against a dataset of Ensembl predicted genes including
mouse and rat genes for out-groups. The resulting pairs were then
filtered and ranked according to genetic distance and gene coverage.
Groups of duplicated human genes were determined by clustering genes
that shared common reciprocal matches. Hence, groups of recently
duplicated genes identified in this manner have phylogenetic meaning
and can be formally defined as being paralogous human genes that have
arisen since the human-rodent divergence. </p>
</li>


</body></html>